// @totvs/thf-core/components/thf-field-base/thf-upload-base/thf-upload-base.service.js
import { Injectable } from '@angular/core';
import { HttpClient, HttpEventType, HttpRequest, HttpResponse } from '@angular/common/http';
var ThfUploadBaseService = (function () {
    /**
     * @param {?} http
     */
    function ThfUploadBaseService(http) {
        this.http = http;
        this.requests = [];
    }
    /**
     * Metódo responsável por enviar os arquivos ao servidor, conforme o parametro URL.
     *
     * @param {?} url URL da requisição a ser efetuada.
     * @param {?} files Arquivos a serem enviados.
     * @param {?} tOnUpload Função a ser executada quando o arquivo for enviado ao servidor.
     * @param {?} uploadCallback Função que será executada enquanto os arquivos estiverem sendo enviados.
     * @param {?} successCallback Função a ser executada quando a requisição for efetuada com sucesso.
     * @param {?} errorCallback Função a ser executada quando a requisição foi efetuada com sucesso.
     * @return {?}
     */
    ThfUploadBaseService.prototype.upload = function (url, files, tOnUpload, uploadCallback, successCallback, errorCallback) {
        var /** @type {?} */ filesLength = files.length;
        var /** @type {?} */ uploadEvent = {
            data: {},
            file: null
        };
        for (var /** @type {?} */ i = 0; i < filesLength; i++) {
            var /** @type {?} */ formData = new FormData();
            var /** @type {?} */ file = files[i];
            formData.append('files', file.rawFile);
            // Função upload, onde o desenvolvedor pode enviar dados para a requisição.
            if (tOnUpload) {
                uploadEvent['file'] = file;
                tOnUpload.emit(uploadEvent);
                formData.append('data', JSON.stringify(uploadEvent.data));
            }
            this.sendFile(url, file, formData, uploadCallback, successCallback, errorCallback);
        }
    };
    /**
     * @param {?} url
     * @param {?} file
     * @param {?} formData
     * @param {?} uploadCallback
     * @param {?} successCallback
     * @param {?} errorCallback
     * @return {?}
     */
    ThfUploadBaseService.prototype.sendFile = function (url, file, formData, uploadCallback, successCallback, errorCallback) {
        var _this = this;
        var /** @type {?} */ request = this.getRequest(url, formData).subscribe(function (event) {
            if (event.type === HttpEventType.UploadProgress) {
                _this.addRequest(file, request);
                var /** @type {?} */ percentDone = Math.round(100 * event.loaded / event.total);
                uploadCallback(file, percentDone);
            }
            else if (event instanceof HttpResponse) {
                // Sucesso, arquivos enviados.
                successCallback(file, event);
            }
        }, function (err) {
            errorCallback(file, err);
        });
    };
    /**
     * @param {?} url
     * @param {?} formData
     * @return {?}
     */
    ThfUploadBaseService.prototype.getRequest = function (url, formData) {
        var /** @type {?} */ req = new HttpRequest('POST', url, formData, {
            reportProgress: true
        });
        return this.http.request(req);
    };
    /**
     * @param {?} file
     * @param {?} callback
     * @return {?}
     */
    ThfUploadBaseService.prototype.stopRequestByFile = function (file, callback) {
        var /** @type {?} */ requestObj = this.requests.find(function (req) {
            return req.file.uid === file.uid;
        });
        if (requestObj) {
            var /** @type {?} */ request = requestObj.request;
            request.unsubscribe();
            this.removeRequest(requestObj);
            callback();
        }
    };
    /**
     * @param {?} requestObj
     * @return {?}
     */
    ThfUploadBaseService.prototype.removeRequest = function (requestObj) {
        var /** @type {?} */ index = this.requests.indexOf(requestObj);
        this.requests.splice(index, 1);
    };
    /**
     * @param {?} file
     * @param {?} request
     * @return {?}
     */
    ThfUploadBaseService.prototype.addRequest = function (file, request) {
        var /** @type {?} */ hasRequest = this.requests.some(function (req) {
            return req.file.uid === file.uid;
        });
        if (!hasRequest) {
            this.requests.push({ file: file, request: request });
        }
    };
    ThfUploadBaseService.decorators = [
        { type: Injectable },
    ];
    /**
     * @nocollapse
     */
    ThfUploadBaseService.ctorParameters = function () { return [
        { type: HttpClient, },
    ]; };
    return ThfUploadBaseService;
}());
export { ThfUploadBaseService };
function ThfUploadBaseService_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfUploadBaseService.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfUploadBaseService.ctorParameters;
    /** @type {?} */
    ThfUploadBaseService.prototype.requests;
    /** @type {?} */
    ThfUploadBaseService.prototype.http;
}
