// @totvs/thf-core/components/thf-field-base/thf-upload-base/thf-upload-file.js
import { ThfUploadStatus } from './thf-upload-status.enum';
var ThfUploadFile = (function () {
    /**
     * @param {?} file
     */
    function ThfUploadFile(file) {
        if (file) {
            this.name = file.name;
            this.extension = this.getExtension(file.name);
            this.size = file.size;
            this.rawFile = file;
            this.uid = this.generateUUID();
            this.status = ThfUploadStatus.None;
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    ThfUploadFile.prototype.getExtension = function (value) {
        if (value) {
            var /** @type {?} */ extension = value.substr(value.lastIndexOf('.'));
            return extension.toLowerCase();
        }
        return '';
    };
    /**
     * @return {?}
     */
    ThfUploadFile.prototype.generateUUID = function () {
        /**
         * @return {?}
         */
        function s4() {
            return Math.floor((1 + Math.random()) * 0x10000)
                .toString(16)
                .substring(1);
        }
        return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
            s4() + '-' + s4() + s4() + s4();
    };
    return ThfUploadFile;
}());
export { ThfUploadFile };
function ThfUploadFile_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfUploadFile.prototype.name;
    /** @type {?} */
    ThfUploadFile.prototype.rawFile;
    /** @type {?} */
    ThfUploadFile.prototype.status;
    /** @type {?} */
    ThfUploadFile.prototype.uid;
    /** @type {?} */
    ThfUploadFile.prototype.extension;
    /** @type {?} */
    ThfUploadFile.prototype.size;
}
