// @totvs/thf-web/components/thf-field/thf-field.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { ThfButtonModule } from '../../components/thf-button/index';
import { ThfCalendarComponent } from '../thf-calendar/thf-calendar.component';
import { ThfCalendarService } from '../thf-calendar/thf-calendar.service';
import { ThfCalendarLangService } from '../thf-calendar/thf-calendar.lang.service';
import { ThfCheckboxGroupComponent } from './thf-checkbox-group/thf-checkbox-group.component';
import { ThfCleanComponent } from './thf-clean/thf-clean.component';
import { ThfDatepickerComponent } from './thf-datepicker/thf-datepicker.component';
import { ThfEmailComponent } from './thf-email/thf-email.component';
import { ThfFieldContainerComponent } from './thf-field-container/thf-field-container.component';
import { ThfInputComponent } from './thf-input/thf-input.component';
import { ThfNumberComponent } from './thf-number/thf-number.component';
import { ThfRadioGroupComponent } from './thf-radio-group/thf-radio-group.component';
import { ThfSelectComponent } from './thf-select/thf-select.component';
import { ThfSwitchComponent } from './thf-switch/thf-switch.component';
import { ThfUploadComponent } from './thf-upload/thf-upload.component';
import { ThfUrlComponent } from './thf-url/thf-url.component';
/**
 * \@description
 *
 * Módulo dos componentes thf-datepicker, thf-email, thf-input, thf-number, thf-radio-group, thf-upload e thf-url
 */
var ThfFieldModule = (function () {
    function ThfFieldModule() {
    }
    ThfFieldModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        HttpClientModule,
                        ThfButtonModule
                    ],
                    exports: [
                        ThfCheckboxGroupComponent,
                        ThfDatepickerComponent,
                        ThfEmailComponent,
                        ThfInputComponent,
                        ThfNumberComponent,
                        ThfRadioGroupComponent,
                        ThfSelectComponent,
                        ThfSwitchComponent,
                        ThfUploadComponent,
                        ThfUrlComponent
                    ],
                    declarations: [
                        ThfCalendarComponent,
                        ThfCheckboxGroupComponent,
                        ThfCleanComponent,
                        ThfDatepickerComponent,
                        ThfEmailComponent,
                        ThfFieldContainerComponent,
                        ThfInputComponent,
                        ThfNumberComponent,
                        ThfRadioGroupComponent,
                        ThfSelectComponent,
                        ThfSwitchComponent,
                        ThfUploadComponent,
                        ThfUrlComponent
                    ],
                    providers: [
                        ThfCalendarService,
                        ThfCalendarLangService
                    ],
                    entryComponents: [
                        ThfCalendarComponent
                    ]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfFieldModule.ctorParameters = function () { return []; };
    return ThfFieldModule;
}());
export { ThfFieldModule };
function ThfFieldModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfFieldModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfFieldModule.ctorParameters;
}
