// @totvs/thf-web/components/thf-list/thf-list.component.js
import { Component, Output, Input, EventEmitter } from '@angular/core';
/**
 * \@description
 *
 * Componente de lista que recebe um array de objetos e renderiza de forma dinâmica os dados de
 * acordo com a necessidade de cada tela. O componente disponibiliza uma área de exibição específica
 * para exibição de mais informações através do link "Exibir Detalhes".
 * Este componente é utilizado em conjunto com os componentes thf-list-item, thf-list-item-content e
 * thf-list-item-detail.
 *
 * \@example
 * <example name='thf-list'>
 * <file name='sample-thf-list.component.html'>
 *   <thf-list>
 *      <thf-list-item *ngFor="let item of data2; let i = index"
 *            [t-title]="item.name">
 *      </thf-list-item>
 *    </thf-list>
 *    <hr/>
 *    <thf-list>
 *      <thf-list-item *ngFor="let item of data2; let i = index"
 *            [t-title]="item.name">
 *            <thf-list-item-content>
 *                {{item.name}}
 *            </thf-list-item-content>
 *      </thf-list-item>
 *    </thf-list>
 *    <hr/>
 *    <thf-list (t-show-more)="showMoreFunction()" [t-show-more-disabled]="showMoreDisabled">
 *        <thf-list-item *ngFor="let item of data; let i = index"
 *          [t-title]="item.title"
 *          [t-index]="i"
 *          [t-actions]="eventsObject">
 *          <thf-list-item-content style="display:flex">
 *            <thf-info t-label="Cadastro" [t-value]="item.register"></thf-info>
 *            <thf-info t-label="E-mail" [t-value]="item.mail" style="margin-left: 80px;"></thf-info>
 *            <thf-info t-label="Última Atualização" [t-value]="item.updated" style="margin-left: 80px;"></thf-info>
 *            <thf-info t-label="Telefone" [t-value]="item.phoneNumber" style="margin-left: 80px;"></thf-info>
 *          </thf-list-item-content>
 *          <thf-list-item-detail>
 *            <div>Outros detalhes a serem exibidos</div>
 *          </thf-list-item-detail>
 *        </thf-list-item>
 *    </thf-list>
 *    <hr/>
 *    <thf-list (t-show-more)="showMoreFunction()" [t-show-more-disabled]="showMoreDisabled">
 *        <thf-list-item *ngFor="let item of data; let i = index"
 *          [t-title]="item.title"
 *          [t-index]="i"
 *          [t-actions]="eventsObject2">
 *          <thf-list-item-content style="display:flex">
 *            <thf-info t-label="Cadastro" [t-value]="item.register"></thf-info>
 *            <thf-info t-label="E-mail" [t-value]="item.mail" style="margin-left: 80px;"></thf-info>
 *            <thf-info t-label="Última Atualização" [t-value]="item.updated" style="margin-left: 80px;"></thf-info>
 *            <thf-info t-label="Telefone" [t-value]="item.phoneNumber" style="margin-left: 80px;"></thf-info>
 *          </thf-list-item-content>
 *          <thf-list-item-detail>
 *            <div>Outros detalhes a serem exibidos</div>
 *          </thf-list-item-detail>
 *       </thf-list-item>
 *   </thf-list>
 * </file>
 *
 * <file name="sample-thf-list.component.ts">
 * import { Component, OnInit } from '\@angular/core';
 *
 * import { ThfModule } from '\@totvs/thf-web';
 *
 * \@Component({
 *   selector: 'sample-thf-list',
 *   templateUrl: './sample-thf-list.component.html',
 * })
 * export class SampleThfListComponent {
 *
 *   showMoreDisabled = false;
 *
 *   data: Array<any> = [
 *     {
 *       title: 'José dos Santos', name: 'José', register: '22/03/2017', mail: 'jose.santos\@email.com.br',
 *       updated: '22/10/2017 as 04:20', phoneNumber: '(47) 8888-8888'
 *     },
 *     {
 *       title: 'Maria da Silva', name: 'Maria', register: '21/03/2017', mail: 'maria.silva\@email.com.br',
 *       updated: '16/10/2017 as 04:20', phoneNumber: '(42) 9888-9999'
 *     }
 *   ];
 *
 *   data2: Array<any> = [
 *     { 'title': 'Marte', 'name': 'Marte' },
 *     { 'title': 'Plutão', 'name': 'Plutão' },
 *     { 'title': 'Terra', 'name': 'Terra' }];
 *
 *   eventsObject: Array<any> = [{ 'label': 'Editar', 'function': 'onClick1' },
 *   { 'label': 'Excluir', 'function': 'onClick2' }];
 *
 *   eventsObject2: Array<any> = [{ 'label': 'Editar', 'function': 'onClick1' },
 *   { 'label': 'Excluir', 'function': 'onClick2' },
 *   { 'label': 'Clonar', 'function': 'onClick3' }];
 *
 *   onClick1() {
 *     alert('onClick1');
 *   }
 *
 *   onClick2() {
 *     alert('onClick2');
 *   }
 *
 *   onClick3() {
 *     alert('onClick3');
 *   }
 *
 *   showMoreFunction() {
 *     this.data.push({
 *       title: 'Marcos Figueira', name: 'Marcos', register: '10/08/2017',
 *       mail: 'marcos.figueira\@email.com.br', updated: '01/08/2017 as 14:28',
 *       phoneNumber: '(12) 9778-9985'
 *     });
 *     this.showMoreDisabled = true;
 *   }
 * }
 *
 * </file>
 * </example>
 */
var ThfListComponent = (function () {
    function ThfListComponent() {
        /**
         * Recebe uma ação de clique do controller da página para o botão "Carregar mais resultados".
         */
        this.showMore = new EventEmitter;
        /**
         * Recebe valores "true" ou "false" para habilitar ou desabilitar o botão "Carregar Mais Resultados"
         */
        this.showMoreDisabled = 'false';
    }
    /**
     * @return {?}
     */
    ThfListComponent.prototype.onShowMore = function () {
        this.showMore.emit(null);
    };
    ThfListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-list', template: "<ng-content></ng-content> <div style=\"width: 100%\" *ngIf=\"showMore.observers.length > 0\"> <div style=\"text-align: center;\"> <thf-button t-label=\"Carregar mais resultados\" (t-click)=\"onShowMore()\" t-id=\"btnShowMoreId\" [t-disabled]=\"showMoreDisabled\"></thf-button> </div> </div>"
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfListComponent.ctorParameters = function () { return []; };
    ThfListComponent.propDecorators = {
        'showMore': [{ type: Output, args: ['t-show-more',] },],
        'showMoreDisabled': [{ type: Input, args: ['t-show-more-disabled',] },],
    };
    return ThfListComponent;
}());
export { ThfListComponent };
function ThfListComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfListComponent.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfListComponent.ctorParameters;
    /** @type {?} */
    ThfListComponent.propDecorators;
    /**
     * Recebe uma ação de clique do controller da página para o botão "Carregar mais resultados".
     * @type {?}
     */
    ThfListComponent.prototype.showMore;
    /**
     * Recebe valores "true" ou "false" para habilitar ou desabilitar o botão "Carregar Mais Resultados"
     * @type {?}
     */
    ThfListComponent.prototype.showMoreDisabled;
}
