// @totvs/thf-web/components/thf-list/thf-list.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThfListComponent } from './thf-list.component';
import { ThfButtonModule } from '../thf-button/thf-button.module';
import { ThfListItemComponent } from './thf-list-item.component';
import { ThfListItemDetailComponent } from './thf-list-item-detail.component';
import { ThfListItemContentComponent } from './thf-list-item-content.component';
/**
 * \@description
 * Módulo do componente thf-list
 */
var ThfListModule = (function () {
    function ThfListModule() {
    }
    ThfListModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        ThfButtonModule
                    ],
                    declarations: [
                        ThfListComponent,
                        ThfListItemComponent,
                        ThfListItemDetailComponent,
                        ThfListItemContentComponent
                    ],
                    exports: [
                        ThfListComponent,
                        ThfListItemComponent,
                        ThfListItemDetailComponent,
                        ThfListItemContentComponent
                    ],
                    providers: [],
                    schemas: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfListModule.ctorParameters = function () { return []; };
    return ThfListModule;
}());
export { ThfListModule };
function ThfListModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfListModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfListModule.ctorParameters;
}
