// @totvs/thf-web/components/thf-info/thf-info.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThfInfoComponent } from './thf-info.component';
/**
 * \@description
 * Módulo do componente thf-info
 */
var ThfInfoModule = (function () {
    function ThfInfoModule() {
    }
    ThfInfoModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        ThfInfoComponent
                    ],
                    exports: [ThfInfoComponent],
                    providers: [],
                    schemas: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfInfoModule.ctorParameters = function () { return []; };
    return ThfInfoModule;
}());
export { ThfInfoModule };
function ThfInfoModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfInfoModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfInfoModule.ctorParameters;
}
