// @totvs/thf-web/components/thf-avatar/thf-avatar.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ThfAvatarComponent } from './thf-avatar.component';
/**
 * \@description
 * Módulo do componente thf-avatar
 */
var ThfAvatarModule = (function () {
    function ThfAvatarModule() {
    }
    ThfAvatarModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        ThfAvatarComponent
                    ],
                    exports: [
                        ThfAvatarComponent
                    ],
                    providers: [],
                    schemas: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfAvatarModule.ctorParameters = function () { return []; };
    return ThfAvatarModule;
}());
export { ThfAvatarModule };
function ThfAvatarModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfAvatarModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfAvatarModule.ctorParameters;
}
