// @totvs/thf-web/components/thf-avatar/thf-avatar.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@totvs/thf-core/components/thf-avatar-base')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/common', '@angular/core', '@totvs/thf-core/components/thf-avatar-base'], factory) :
	(factory((global['thf-avatar'] = {}),global.common,global.core,global.thfAvatarBase));
}(this, (function (exports,common,core,thfAvatarBase) { 'use strict';

var __extends = (undefined && undefined.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 *
 * \@example
 * <example name="thf-avatar-labs" title="Totvs Avatar Labs" >
 *  <file name='sample-thf-avatar-labs.component.html'> </file>
 *  <file name="sample-thf-avatar-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-avatar" title="Totvs Avatar com diferentes imagens" >
 *  <file name='sample-thf-avatar.component.html'> </file>
 *  <file name="sample-thf-avatar.component.ts"> </file>
 * </example>
 */
var ThfAvatarComponent = (function (_super) {
    __extends(ThfAvatarComponent, _super);
    function ThfAvatarComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.defaultSrc = './assets/images/totvs-avatar-default.svg';
        return _this;
    }
    /**
     * @return {?}
     */
    ThfAvatarComponent.prototype.ngOnInit = function () {
        if (!this.src) {
            this.src = this.defaultSrc;
        }
    };
    /**
     * @return {?}
     */
    ThfAvatarComponent.prototype.onError = function () {
        this.src = this.defaultSrc;
    };
    ThfAvatarComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'thf-avatar', template: "<div class=\"thf-avatar thf-avatar-{{size}}\"> <img [src]=\"src\" alt=\"\" (error)=\"onError()\" class=\"thf-avatar-image\" > </div> "
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfAvatarComponent.ctorParameters = function () { return []; };
    return ThfAvatarComponent;
}(thfAvatarBase.ThfAvatarBaseComponent));

/**
 * \@description
 * Módulo do componente thf-avatar
 */
var ThfAvatarModule = (function () {
    function ThfAvatarModule() {
    }
    ThfAvatarModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule
                    ],
                    declarations: [
                        ThfAvatarComponent
                    ],
                    exports: [
                        ThfAvatarComponent
                    ],
                    providers: [],
                    schemas: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfAvatarModule.ctorParameters = function () { return []; };
    return ThfAvatarModule;
}());

exports.ThfAvatarModule = ThfAvatarModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
