// @totvs/thf-web/components/thf-avatar/thf-avatar.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component } from '@angular/core';
import { ThfAvatarBaseComponent } from '@totvs/thf-core/components/thf-avatar-base';
/**
 *
 * \@example
 * <example name="thf-avatar-labs" title="Totvs Avatar Labs" >
 *  <file name='sample-thf-avatar-labs.component.html'> </file>
 *  <file name="sample-thf-avatar-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-avatar" title="Totvs Avatar com diferentes imagens" >
 *  <file name='sample-thf-avatar.component.html'> </file>
 *  <file name="sample-thf-avatar.component.ts"> </file>
 * </example>
 */
var ThfAvatarComponent = (function (_super) {
    __extends(ThfAvatarComponent, _super);
    function ThfAvatarComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.defaultSrc = './assets/images/totvs-avatar-default.svg';
        return _this;
    }
    /**
     * @return {?}
     */
    ThfAvatarComponent.prototype.ngOnInit = function () {
        if (!this.src) {
            this.src = this.defaultSrc;
        }
    };
    /**
     * @return {?}
     */
    ThfAvatarComponent.prototype.onError = function () {
        this.src = this.defaultSrc;
    };
    ThfAvatarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-avatar', template: "<div class=\"thf-avatar thf-avatar-{{size}}\"> <img [src]=\"src\" alt=\"\" (error)=\"onError()\" class=\"thf-avatar-image\" > </div> "
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfAvatarComponent.ctorParameters = function () { return []; };
    return ThfAvatarComponent;
}(ThfAvatarBaseComponent));
export { ThfAvatarComponent };
function ThfAvatarComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfAvatarComponent.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfAvatarComponent.ctorParameters;
    /** @type {?} */
    ThfAvatarComponent.prototype.defaultSrc;
}
