// @totvs/thf-web/services/thf-dialog/thf-dialog.module.js
import { NgModule } from '@angular/core';
import { ThfModalModule } from '../../components/thf-modal/thf-modal.module';
import { ThfDialogComponent } from './thf-dialog.component';
import { ThfDialogService } from './thf-dialog.service';
var ThfDialogModule = (function () {
    function ThfDialogModule() {
    }
    ThfDialogModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        ThfDialogComponent
                    ],
                    imports: [
                        ThfModalModule
                    ],
                    exports: [
                        ThfDialogComponent
                    ],
                    entryComponents: [
                        ThfDialogComponent
                    ],
                    providers: [
                        ThfDialogService
                    ],
                    bootstrap: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfDialogModule.ctorParameters = function () { return []; };
    return ThfDialogModule;
}());
export { ThfDialogModule };
function ThfDialogModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfDialogModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfDialogModule.ctorParameters;
}
