// @totvs/thf-web/components/thf-modal/thf-modal.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThfButtonModule } from './../thf-button/thf-button.module';
import { ThfModalComponent } from './thf-modal.component';
/**
 * \@description
 * Módulo do componente thf-modal
 */
var ThfModalModule = (function () {
    function ThfModalModule() {
    }
    ThfModalModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        ThfButtonModule
                    ],
                    declarations: [
                        ThfModalComponent
                    ],
                    exports: [ThfModalComponent],
                    providers: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfModalModule.ctorParameters = function () { return []; };
    return ThfModalModule;
}());
export { ThfModalModule };
function ThfModalModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfModalModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfModalModule.ctorParameters;
}
