// @totvs/thf-web/services/thf-i18n/thf-i18n.module.js
import { NgModule, InjectionToken } from '@angular/core';
import { Http } from '@angular/http';
import { ThfI18nService, returnThfI18nService } from './thf-i18n.service';
/**
 * \@description
 * Módulo para controle de idiomas com THF.
 *
 * Para utilização do serviço de idiomas **thfI18nService**, deve-se importar este módulo mesmo já tendo importado
 * o módulo ThfModule.
 * Na importação deve ser invocado o método **config** informando um objeto para configuração.
 *
 * Exemplo de configuração:
 * ```
 * import { ThfI18nConfig } from '\@totvs/thf-core/services/thf-i18n-base/thf-i18n-config.interface';
 *
 * const i18nConfig: ThfI18nConfig = {
 *   default: {
 *     language: 'pt-BR',
 *     context: 'general',
 *     cache: true
 *   },
 *   contexts: {
 *     general: {
 *       'pt-BR': GeneralPT,
 *       'en-US': GeneralEN
 *     },
 *     hcm: {
 *       url: 'http://10.1.1.1/api/translations/hcm/'
 *     }
 *   }
 * };
 *
 * \@NgModule({
 *   declarations: [],
 *   imports: [
 *     ThfModule,
 *     ThfI18nModule.config(i18nConfig)
 *   ],
 *   bootstrap: [AppComponent]
 * })
 * ```
 */
var I18N_CONFIG = new InjectionToken('');
var ThfI18nModule = (function () {
    function ThfI18nModule() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    ThfI18nModule.config = function (config) {
        return {
            ngModule: ThfI18nModule,
            providers: [
                {
                    provide: I18N_CONFIG,
                    useValue: config
                },
                {
                    provide: ThfI18nService,
                    useFactory: returnThfI18nService,
                    deps: [I18N_CONFIG, Http]
                }
            ]
        };
    };
    ThfI18nModule.decorators = [
        { type: NgModule, args: [{},] },
    ];
    /**
     * @nocollapse
     */
    ThfI18nModule.ctorParameters = function () { return []; };
    return ThfI18nModule;
}());
export { ThfI18nModule };
function ThfI18nModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfI18nModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfI18nModule.ctorParameters;
}
