// @totvs/thf-web/services/thf-component-injector/thf-component-injector.service.js
import { Injectable, ComponentFactoryResolver, ApplicationRef, Injector } from '@angular/core';
var ThfComponentInjectorService = (function () {
    /**
     * @param {?} componentResolver
     * @param {?} applicationRef
     * @param {?} injector
     */
    function ThfComponentInjectorService(componentResolver, applicationRef, injector) {
        this.componentResolver = componentResolver;
        this.applicationRef = applicationRef;
        this.injector = injector;
    }
    /**
     * @param {?} component
     * @return {?}
     */
    ThfComponentInjectorService.prototype.createComponentInApplication = function (component) {
        var /** @type {?} */ componentRef = this.componentResolver.resolveComponentFactory(component).create(this.injector);
        this.applicationRef.attachView(componentRef.hostView);
        var /** @type {?} */ domElem = (((componentRef.hostView)).rootNodes[0]);
        document.body.appendChild(domElem);
        return componentRef;
    };
    /**
     * @param {?} componentRef
     * @return {?}
     */
    ThfComponentInjectorService.prototype.destroyComponentInApplication = function (componentRef) {
        this.applicationRef.detachView(componentRef.hostView);
        componentRef.destroy();
    };
    ThfComponentInjectorService.decorators = [
        { type: Injectable },
    ];
    /**
     * @nocollapse
     */
    ThfComponentInjectorService.ctorParameters = function () { return [
        { type: ComponentFactoryResolver, },
        { type: ApplicationRef, },
        { type: Injector, },
    ]; };
    return ThfComponentInjectorService;
}());
export { ThfComponentInjectorService };
function ThfComponentInjectorService_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfComponentInjectorService.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfComponentInjectorService.ctorParameters;
    /** @type {?} */
    ThfComponentInjectorService.prototype.componentResolver;
    /** @type {?} */
    ThfComponentInjectorService.prototype.applicationRef;
    /** @type {?} */
    ThfComponentInjectorService.prototype.injector;
}
