// @totvs/thf-web/services/thf-component-injector/thf-component-injector.module.js
import { NgModule } from '@angular/core';
import { ThfComponentInjectorService } from './thf-component-injector.service';
var ThfComponentInjectorModule = (function () {
    function ThfComponentInjectorModule() {
    }
    ThfComponentInjectorModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [],
                    exports: [],
                    providers: [
                        ThfComponentInjectorService
                    ],
                    bootstrap: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfComponentInjectorModule.ctorParameters = function () { return []; };
    return ThfComponentInjectorModule;
}());
export { ThfComponentInjectorModule };
function ThfComponentInjectorModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfComponentInjectorModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfComponentInjectorModule.ctorParameters;
}
