// @totvs/thf-web/services/services.module.js
import { NgModule } from '@angular/core';
import { ThfComponentInjectorModule } from './thf-component-injector/thf-component-injector.module';
import { ThfDialogModule } from './thf-dialog/thf-dialog.module';
import { ThfNotificationModule } from './thf-notification/thf-notification.module';
import { ThfI18nPipe } from './thf-i18n/thf-i18n.pipe';
var ThfServicesModule = (function () {
    function ThfServicesModule() {
    }
    ThfServicesModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        ThfI18nPipe
                    ],
                    imports: [
                        ThfComponentInjectorModule,
                        ThfDialogModule,
                        ThfNotificationModule
                    ],
                    exports: [
                        ThfComponentInjectorModule,
                        ThfDialogModule,
                        ThfNotificationModule,
                        ThfI18nPipe
                    ],
                    providers: [],
                    bootstrap: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfServicesModule.ctorParameters = function () { return []; };
    return ThfServicesModule;
}());
export { ThfServicesModule };
function ThfServicesModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfServicesModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfServicesModule.ctorParameters;
}
