// @totvs/thf-web/services/thf-notification/thf-notification.module.js
import { NgModule } from '@angular/core';
import { ThfToasterComponent } from './thf-toaster/thf-toaster.component';
import { CommonModule } from '@angular/common';
var ThfNotificationModule = (function () {
    function ThfNotificationModule() {
    }
    ThfNotificationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        ThfToasterComponent
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [],
                    entryComponents: [
                        ThfToasterComponent
                    ],
                    providers: [],
                    bootstrap: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfNotificationModule.ctorParameters = function () { return []; };
    return ThfNotificationModule;
}());
export { ThfNotificationModule };
function ThfNotificationModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfNotificationModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfNotificationModule.ctorParameters;
}
