// @totvs/thf-web/components/thf-tabs/thf-tabs.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@totvs/thf-core/components/thf-tabs-base')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', '@totvs/thf-core/components/thf-tabs-base'], factory) :
	(factory((global['thf-tabs'] = {}),global.core,global.common,global.thfTabsBase));
}(this, (function (exports,core,common,thfTabsBase) { 'use strict';

var ThfTabButtonComponent = (function () {
    /**
     * @param {?} el
     */
    function ThfTabButtonComponent(el) {
        this.el = el;
        // Método recebido do usuário para ser disparado quando clicar na aba
        this.clickCallback = new core.EventEmitter;
        this.clickTab = new core.EventEmitter;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    ThfTabButtonComponent.prototype.onClick = function (e) {
        if (!this.disabled) {
            this.clickTab.emit(this.id);
            this.clickCallback.emit(this.id);
        }
    };
    /**
     * @return {?}
     */
    ThfTabButtonComponent.prototype.controlDisabled = function () {
        if (this.disabled) {
            this.el.nativeElement.classList.add('thf-tab-disabled');
        }
        else {
            this.el.nativeElement.classList.remove('thf-tab-disabled');
        }
    };
    /**
     * @return {?}
     */
    ThfTabButtonComponent.prototype.controlActive = function () {
        if (this.active) {
            this.el.nativeElement.classList.add('thf-tab-active');
        }
        else {
            this.el.nativeElement.classList.remove('thf-tab-active');
        }
    };
    /**
     * @return {?}
     */
    ThfTabButtonComponent.prototype.ngAfterViewChecked = function () {
        if (this.hide) {
            this.el.nativeElement.style.display = 'none';
        }
        else {
            this.el.nativeElement.style.display = '';
            this.el.nativeElement.style.width = this.width + '%';
        }
        this.controlDisabled();
        this.controlActive();
    };
    ThfTabButtonComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'thf-tab-button', template: "<div #tab [ngClass]=\"{'thf-tab-size1': !small, 'thf-tab-size2': small}\"> {{label}} </div> ", styles: [":host(thf-tab-button) { background-color: red; display: table-cell; background-color: transparent; box-shadow: inset 0 -1px 0 0 #edefef; text-align: center; vertical-align: middle; font-family: NunitoSans; font-weight: bold; color: #0c9abe; cursor: pointer; } :host(thf-tab-button) .thf-tab-size1 { display: inline-flex; align-items: center; text-align: center; height: 44px; font-size: 16px; } :host(thf-tab-button) .thf-tab-size2 { display: inline-flex; align-items: center; text-align: center; height: 32px; font-size: 14px; } :host(thf-tab-button) .thf-tab-hidden { display: none; } :host(thf-tab-button).thf-tab-disabled { color: rgba(0, 0, 0, 0.4); cursor: initial; } :host(thf-tab-button).thf-tab-active { box-shadow: inset 0 -3px 0 0 #29b6c5; } "]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfTabButtonComponent.ctorParameters = function () { return [
        { type: core.ElementRef, },
    ]; };
    ThfTabButtonComponent.propDecorators = {
        'id': [{ type: core.Input },],
        'active': [{ type: core.Input },],
        'label': [{ type: core.Input },],
        'width': [{ type: core.Input },],
        'small': [{ type: core.Input },],
        'disabled': [{ type: core.Input },],
        'hide': [{ type: core.Input },],
        'clickCallback': [{ type: core.Output },],
        'clickTab': [{ type: core.Output },],
        'tab': [{ type: core.ViewChild, args: ['tab',] },],
        'onClick': [{ type: core.HostListener, args: ['click', ['$event'],] },],
    };
    return ThfTabButtonComponent;
}());

var __extends$1 = (undefined && undefined.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var ThfTabComponent = (function (_super) {
    __extends$1(ThfTabComponent, _super);
    /**
     * @param {?} el
     */
    function ThfTabComponent(el) {
        var _this = _super.call(this) || this;
        _this.el = el;
        return _this;
    }
    /**
     * @return {?}
     */
    ThfTabComponent.prototype.ngAfterViewChecked = function () {
        this.el.nativeElement.style.display = (this.active) ? '' : 'none';
    };
    ThfTabComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'thf-tab', template: "<div id=\"content_tab_{{id}}\"> <ng-content></ng-content> </div> "
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfTabComponent.ctorParameters = function () { return [
        { type: core.ElementRef, },
    ]; };
    return ThfTabComponent;
}(thfTabsBase.ThfTabBaseComponent));

var __extends = (undefined && undefined.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 *
 * \@example
 *  <example name='thf-tabs'>
 *  <file name='sample-thf-tabs.component.html'>
 *
 *    <thf-tabs>
 *      <thf-tab t-label="Aba 1">
 *        <h1>Conteúdo da aba 1</h1>
 *      </thf-tab>
 *      <thf-tab t-label="Aba 2" (t-click)="metodo()">
 *        <h1>Conteúdo da aba 2</h1>
 *      </thf-tab>
 *      <thf-tab t-label="Aba 3" t-active>
 *        <h1>Conteúdo da aba 3</h1>
 *      </thf-tab>
 *    </thf-tabs>
 *
 *    <thf-tabs t-small>
 *      <thf-tab t-label="Aba 1" t-active>
 *        <h1>Conteúdo da aba 1</h1>
 *      </thf-tab>
 *      <thf-tab t-label="Aba 2" [t-hide]="hide">
 *        <h1>Conteúdo da aba 2</h1>
 *      </thf-tab>
 *      <thf-tab t-label="Aba 3" t-disabled>
 *        <h1>Conteúdo da aba 3</h1>
 *      </thf-tab>
 *      <thf-tab t-label="Aba 4">
 *        <h1>Conteúdo da aba 4</h1>
 *      </thf-tab>
 *    </thf-tabs>
 *
 *    <thf-button (click)="hide=!hide" [t-label]="'hide: '+hide"></thf-button>
 *
 *  </file>
 *  <file name='sample-thf-tabs.component.ts'>
 *   import { Component, AfterViewChecked } from '\@angular/core';
 *   import { ThfModule } from '\@totvs/thf-web';
 *
 *   \@Component({
 *     selector: 'sample-thf-tabs',
 *     templateUrl: './sample-thf-tabs.component.html',
 *   })
 *   export class SampleThfTabsComponent {
 *
 *     hide: boolean = false;
 *
 *     metodo() {
 *       console.log('Disparado método ao clicar na aba');
 *     }
 *
 *   }
 *  </file>
 * </example>
 */
var ThfTabsComponent = (function (_super) {
    __extends(ThfTabsComponent, _super);
    /**
     * @param {?} resolver
     */
    function ThfTabsComponent(resolver) {
        var _this = _super.call(this) || this;
        _this.resolver = resolver;
        // Id do componente
        _this.listButtons = [];
        _this.created = false;
        return _this;
    }
    /**
     * @return {?}
     */
    ThfTabsComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        var /** @type {?} */ ComponentFactory = this.resolver.resolveComponentFactory(ThfTabButtonComponent);
        // Calcula a largura das tabs
        var /** @type {?} */ width = this.calculateWidth(this.countVisibledTabs());
        // Percorre tabs
        this.tabsChild.forEach(function (tab) {
            // Cria os botões
            var /** @type {?} */ compButton = _this.nav.createComponent(ComponentFactory);
            compButton.instance.label = tab.label;
            compButton.instance.id = tab.id;
            compButton.instance.active = tab.active;
            compButton.instance.hide = tab.hide;
            compButton.instance.disabled = tab.disabled;
            compButton.instance.width = width;
            compButton.instance.small = _this.small;
            compButton.instance.clickCallback = tab.click;
            // É necessário esta arrow function para não perder o contexto.
            // Caso contrário, a funcão não teria mais acesso ao "this" deste component
            compButton.instance.clickTab.subscribe(function (id) { return _this.selectedTab(id); });
            _this.listButtons.push(compButton.instance);
        });
        this.created = true;
    };
    /**
     * @return {?}
     */
    ThfTabsComponent.prototype.ngAfterViewChecked = function () {
        var _this = this;
        if (this.created) {
            // Váriaveis utilizadas para trocar aba ativa,
            // quando a aba ativa está desabilitada ou escondida
            var /** @type {?} */ changeTab_1 = false;
            var /** @type {?} */ lastEnabledTab_1 = null;
            var /** @type {?} */ lastEnabledButton_1 = null;
            // Calcula a largura das tabs
            var /** @type {?} */ width_1 = this.calculateWidth(this.countVisibledTabs());
            this.tabsChild.forEach(function (tab) {
                _this.listButtons.forEach(function (button) {
                    if (button.id === tab.id) {
                        // Atualiza a largura dos botões de acordo com os botões visíveis
                        button.width = width_1;
                        button.active = tab.active;
                        if (tab.disabled || tab.hide) {
                            // Verifica se uma desativada ou escondida está ativa
                            if (tab.active) {
                                tab.active = false;
                                button.active = false;
                                changeTab_1 = true;
                            }
                        }
                        else {
                            // Armazena a aba que não esteja escondida para trocar de aba automaticamente
                            // caso a aba ativa seja escondida ou desabilitada.
                            if (!changeTab_1) {
                                lastEnabledTab_1 = tab;
                                lastEnabledButton_1 = button;
                            }
                            else if (!lastEnabledTab_1) {
                                lastEnabledTab_1 = tab;
                                lastEnabledButton_1 = button;
                            }
                        }
                        // Atualiza as propriedades disabled e hide das abas
                        button.disabled = tab.disabled;
                        button.hide = tab.hide;
                    }
                });
            });
            // Habilita uma aba ativa
            if (changeTab_1 && lastEnabledTab_1) {
                lastEnabledTab_1.active = true;
                lastEnabledButton_1.active = true;
            }
        }
    };
    /**
     * @param {?} id
     * @return {?}
     */
    ThfTabsComponent.prototype.selectedTab = function (id) {
        this.tabsChild.forEach(function (tab) {
            tab.active = (tab.id === id);
        });
        this.listButtons.forEach(function (button) {
            if (button.id === id) {
                button.active = true;
            }
            else {
                button.active = false;
            }
        });
    };
    /**
     * @return {?}
     */
    ThfTabsComponent.prototype.countVisibledTabs = function () {
        var /** @type {?} */ cont = 0;
        this.tabsChild.forEach(function (tab) {
            cont = (tab.hide) ? cont : cont + 1;
        });
        return cont;
    };
    /**
     * @param {?} visibledTabs
     * @return {?}
     */
    ThfTabsComponent.prototype.calculateWidth = function (visibledTabs) {
        var /** @type {?} */ width = parseFloat((100 / visibledTabs).toFixed(2));
        if (width * visibledTabs > 100) {
            width = parseFloat(String(width).substring(0, String(width).length - 1));
        }
        return width;
    };
    ThfTabsComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'thf-tabs', template: "<div class=\"thf-tabs-parent\"> <div #nav></div> </div> <div class=\"thf-tabs-container\"> <ng-content></ng-content> </div> ", styles: [".thf-tabs-parent { width: 100%; display: table; } .thf-tabs-container { padding-top: 5px; } "]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfTabsComponent.ctorParameters = function () { return [
        { type: core.ComponentFactoryResolver, },
    ]; };
    ThfTabsComponent.propDecorators = {
        'tabsChild': [{ type: core.ContentChildren, args: [ThfTabComponent,] },],
        'nav': [{ type: core.ViewChild, args: ['nav', { read: core.ViewContainerRef },] },],
    };
    return ThfTabsComponent;
}(thfTabsBase.ThfTabsBaseComponent));

/**
 * \@description
 * Módulo do componente thf-tabs
 */
var ThfTabsModule = (function () {
    function ThfTabsModule() {
    }
    ThfTabsModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule
                    ],
                    declarations: [
                        ThfTabsComponent,
                        ThfTabComponent,
                        ThfTabButtonComponent
                    ],
                    entryComponents: [
                        ThfTabButtonComponent
                    ],
                    exports: [
                        ThfTabsComponent,
                        ThfTabComponent
                    ]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfTabsModule.ctorParameters = function () { return []; };
    return ThfTabsModule;
}());

exports.ThfTabsModule = ThfTabsModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
