// @totvs/thf-web/components/thf-tabs/thf-tabs.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThfTabsComponent } from './thf-tabs.component';
import { ThfTabComponent } from './thf-tab/thf-tab.component';
import { ThfTabButtonComponent } from './thf-tab-button/thf-tab-button.component';
/**
 * \@description
 * Módulo do componente thf-tabs
 */
var ThfTabsModule = (function () {
    function ThfTabsModule() {
    }
    ThfTabsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        ThfTabsComponent,
                        ThfTabComponent,
                        ThfTabButtonComponent
                    ],
                    entryComponents: [
                        ThfTabButtonComponent
                    ],
                    exports: [
                        ThfTabsComponent,
                        ThfTabComponent
                    ]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfTabsModule.ctorParameters = function () { return []; };
    return ThfTabsModule;
}());
export { ThfTabsModule };
function ThfTabsModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfTabsModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfTabsModule.ctorParameters;
}
