// @totvs/thf-web/components/thf-widget/thf-widget.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@totvs/thf-core/components/thf-widget-base')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', '@totvs/thf-core/components/thf-widget-base'], factory) :
	(factory((global['thf-widget'] = {}),global.core,global.common,global.thfWidgetBase));
}(this, (function (exports,core,common,thfWidgetBase) { 'use strict';

var __extends = (undefined && undefined.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 *
 * \@example
 * <example name='thf-widget-labs' title='Totvs Widget Labs'>
 *  <file name='sample-thf-widget-labs.component.html'> </file>
 *  <file name='sample-thf-widget-labs.component.ts'> </file>
 * </example>
 *
 * <example name='thf-widget-chart' title='Mostrando gráficos'>
 *  <file name='sample-thf-widget-chart.component.html'> </file>
 *  <file name='sample-thf-widget-chart.component.ts'> </file>
 * </example>
 *
 * <example name='thf-widget' title='Trocando a visualização'>
 *  <file name='sample-thf-widget.component.html'> </file>
 *  <file name='sample-thf-widget.component.ts'> </file>
 * </example>
 *
 * <example name='thf-widget-background' title='Utilizando imagem de background'>
 *  <file name='sample-thf-widget-background.component.html'> </file>
 *  <file name='sample-thf-widget-background.component.ts'> </file>
 * </example>
 */
var ThfWidgetComponent = (function (_super) {
    __extends(ThfWidgetComponent, _super);
    /**
     * @param {?} viewRef
     */
    function ThfWidgetComponent(viewRef) {
        var _this = _super.call(this) || this;
        /**
         * Função chamada ao clicar no ícone de configuração
         */
        _this.setting = new core.EventEmitter;
        // Obtém a referência ao componente pai
        _this.parentRef = viewRef['_view']['component'];
        return _this;
    }
    /**
     * @return {?}
     */
    ThfWidgetComponent.prototype.ngOnInit = function () {
        if (this.background) {
            // Seta uma imagem de background e um tamanho mínimo para que o Widget apareça
            this.elBackground.nativeElement.style.backgroundImage = "url('" + this.background + "')";
            this.elBackground.nativeElement.style.minHeight = '20px';
        }
        if (this.height) {
            var /** @type {?} */ heightBody = this.height - 32; // 32 de padding
            // Calcula tamanho restante para o Body, subtraindo title e footer se existirem
            if (this.title) {
                heightBody = heightBody - 40;
            }
            if (this.primaryLabel) {
                heightBody = heightBody - 40;
            }
            // Seta o corpo do widget com o tamanho calculado
            this.elBody.nativeElement.style.height = heightBody + "px";
        }
    };
    /**
     * @return {?}
     */
    ThfWidgetComponent.prototype.runPrimaryAction = function () {
        this.primaryAction.emit();
    };
    /**
     * @return {?}
     */
    ThfWidgetComponent.prototype.runSecundaryAction = function () {
        this.secundaryAction.emit();
    };
    /**
     * @return {?}
     */
    ThfWidgetComponent.prototype.settingOutput = function () {
        this.setting.emit();
    };
    /**
     * @return {?}
     */
    ThfWidgetComponent.prototype.openHelp = function () {
        window.open(this.help, '_blank');
    };
    /**
     * @param {?} object
     * @param {?} type
     * @return {?}
     */
    ThfWidgetComponent.prototype.isTypeof = function (object, type) {
        return typeof object === type;
    };
    ThfWidgetComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'thf-widget', template: "<div class=\"thf-widget thf-widget-{{ size }}\" #elemBackground> <div *ngIf=\"!isTypeof(title, 'undefined')\" class=\"thf-widget-header-title\"> {{title}} <div *ngIf=\"setting.observers[0] || help\" class=\"thf-pull-right\"> <span id=\"settingAction\" *ngIf=\"setting.observers[0]\" class=\"thf-clickable thf-icon thf-icon-settings\" (click)=\"settingOutput()\"></span> <span id=\"helpLink\" *ngIf=\"help\" class=\"thf-clickable glyphicon glyphicon-question-sign\" (click)=\"openHelp()\"></span> </div> </div> <div class=\"thf-widget-body\" #elemBody> <ng-content></ng-content> </div> <div *ngIf=\"!isTypeof(primaryLabel, 'undefined')\" class=\"thf-widget-footer\"> <div class=\"thf-widget-xl\" *ngIf=\"isTypeof(secundaryLabel, 'undefined')\"> <a class=\"thf-widget-action\" id='primaryAct' (click)=\"runPrimaryAction()\">{{ primaryLabel }}</a> </div> <div class=\"thf-widget-md\" *ngIf=\"!isTypeof(secundaryLabel, 'undefined')\"> <a class=\"thf-widget-action\" id='primaryAct' (click)=\"runPrimaryAction()\">{{ primaryLabel }}</a> </div> <div class=\"thf-widget-md\" *ngIf=\"!isTypeof(secundaryLabel, 'undefined')\"> <a class=\"thf-widget-action\" id='secundaryAct' (click)=\"runSecundaryAction()\">{{ secundaryLabel }}</a> </div> </div> </div>"
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfWidgetComponent.ctorParameters = function () { return [
        { type: core.ViewContainerRef, },
    ]; };
    ThfWidgetComponent.propDecorators = {
        'setting': [{ type: core.Output, args: ['t-setting',] },],
        'elBackground': [{ type: core.ViewChild, args: ['elemBackground', { read: core.ElementRef },] },],
        'elBody': [{ type: core.ViewChild, args: ['elemBody', { read: core.ElementRef },] },],
    };
    return ThfWidgetComponent;
}(thfWidgetBase.ThfWidgetBaseComponent));

/**
 * \@description
 * Módulo do componente thf-widget
 */
var ThfWidgetModule = (function () {
    function ThfWidgetModule() {
    }
    ThfWidgetModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule
                    ],
                    exports: [
                        ThfWidgetComponent
                    ],
                    declarations: [ThfWidgetComponent]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfWidgetModule.ctorParameters = function () { return []; };
    return ThfWidgetModule;
}());

exports.ThfWidgetModule = ThfWidgetModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
