// @totvs/thf-web/components/thf-grid/thf-grid.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ExcelModule, GridModule, PDFModule } from '@progress/kendo-angular-grid';
import { ThfButtonModule } from '../thf-button/thf-button.module';
import { ThfGridComponent } from './thf-grid.component';
import { ThfModalModule } from '../thf-modal/thf-modal.module';
import { ThfRowClickDirective } from './thf-row-click.directive';
/**
 * \@description
 * Módulo do componente thf-grid
 */
var ThfGridModule = (function () {
    function ThfGridModule() {
    }
    ThfGridModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        GridModule,
                        ExcelModule,
                        PDFModule,
                        ThfButtonModule,
                        ThfModalModule
                    ],
                    declarations: [
                        ThfGridComponent,
                        ThfRowClickDirective
                    ],
                    exports: [ThfGridComponent],
                    providers: [],
                    schemas: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfGridModule.ctorParameters = function () { return []; };
    return ThfGridModule;
}());
export { ThfGridModule };
function ThfGridModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfGridModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfGridModule.ctorParameters;
}
