// @totvs/thf-web/components/thf-grid/thf-grid.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@progress/kendo-angular-grid'), require('@totvs/thf-core/components/thf-button-base'), require('@angular/forms'), require('@progress/kendo-data-query'), require('@totvs/thf-core/components/thf-grid-base/thf-grid-base.component'), require('@totvs/thf-core/components/thf-modal-base')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', '@progress/kendo-angular-grid', '@totvs/thf-core/components/thf-button-base', '@angular/forms', '@progress/kendo-data-query', '@totvs/thf-core/components/thf-grid-base/thf-grid-base.component', '@totvs/thf-core/components/thf-modal-base'], factory) :
	(factory((global['thf-grid'] = {}),global.core,global.common,global.kendoAngularGrid,global.thfButtonBase,global.forms,global.kendoDataQuery,global.thfGridBase_component,global.thfModalBase));
}(this, (function (exports,core,common,kendoAngularGrid,thfButtonBase,forms,kendoDataQuery,thfGridBase_component,thfModalBase) { 'use strict';

var __extends = (undefined && undefined.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 *
 * \@example
 * <example name='thf-button-labs' title='Totvs Button Labs'>
 *  <file name='sample-thf-button-labs.component.html'> </file>
 *  <file name='sample-thf-button-labs.component.ts'> </file>
 * </example>
 *
 * <example name='thf-button-click' title='Eventos de click e botão desabilitado'>
 *  <file name='sample-thf-button-click.component.html'> </file>
 *  <file name='sample-thf-button-click.component.ts'> </file>
 *  <file name='sample-thf-button-click.component.css'> </file>
 * </example>
 *
 * <example name='thf-button' title='Tipos e tamanhos do thf-button' >
 *  <file name='sample-thf-button.component.html'> </file>
 * <file name='sample-thf-button.component.ts'> </file>
 * </example>
 */
var ThfButtonComponent = (function (_super) {
    __extends(ThfButtonComponent, _super);
    function ThfButtonComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ThfButtonComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'thf-button', template: "<button class=\"thf-button\" type=\"button\" id=\"{{ id }}\" [disabled]=\"disabled\" [ngClass]=\"{ 'thf-button--primary' : primary, 'thf-button--sm' : small }\" (click)=\"onClick()\"> <span *ngIf=\"icon\" class=\"thf-icon {{ icon }}\" aria-hidden=\"true\"></span> {{ label }} </button> ",
                    providers: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfButtonComponent.ctorParameters = function () { return []; };
    return ThfButtonComponent;
}(thfButtonBase.ThfButtonBaseComponent));

/**
 * \@description
 * Módulo do componente thf-button.
 */
var ThfButtonModule = (function () {
    function ThfButtonModule() {
    }
    ThfButtonModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule
                    ],
                    declarations: [ThfButtonComponent],
                    exports: [ThfButtonComponent]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfButtonModule.ctorParameters = function () { return []; };
    return ThfButtonModule;
}());

var __extends$1 = (undefined && undefined.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 *
 * \@description
 *
 * Para o correto funcionamento do THF Grid deve ser importado o módulo `BrowserAnimationsModule` no módulo principal da sua aplicação.
 *
 * > Módulo da sua aplicação:
 *
 * ```
 * import { BrowserAnimationsModule } from '\@angular/platform-browser/animations';
 *
 * ...
 *
 * \@NgModule({
 *   imports: [
 *     BrowserModule,
 *     BrowserAnimationsModule,
 *     ...
 *     ThfModule,
 *     ...
 *   ],
 *   declarations: [
 *     AppComponent
 *   ],
 *   providers: [],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule { }
 * ```
 *
 * No arquivo `angular-cli.json` da sua aplicação deve ser importado o arquivo `all.css`, que pertence ao
 * pacote `\@progress/kendo-theme-default`.
 *
 * > Arquivo .angular-cli.json:
 *
 * ```
 * ...
 * "styles": [
 *   "../node_modules/\@progress/kendo-theme-default/dist/all.css"
 * ]
 * ...
 * ```
 *
 * \@example
 * <example name="thf-grid" title="Totvs Grid Labs">
 *  <file name='sample-thf-grid.component.html'> </file>
 *  <file name='sample-thf-grid.component.ts'> </file>
 * </example>
 */
var ThfGridComponent = (function (_super) {
    __extends$1(ThfGridComponent, _super);
    /**
     * @param {?} viewRef
     * @param {?} renderer
     */
    function ThfGridComponent(viewRef, renderer) {
        var _this = _super.call(this) || this;
        _this.viewRef = viewRef;
        _this.renderer = renderer;
        /**
         * Habilita a opção de ordenação dos dados nas colunas.
         */
        _this.sortable = false;
        /**
         * Habilita a opção de selecionar uma linha do grid
         */
        _this.selectable = false;
        /**
         * Habilita o botão para edição da coluna.
         */
        _this.editable = false;
        /**
         * Habilita a opção para agrupamento permitindo agrupar no máximo dois níveis.
         */
        _this.groupable = false;
        /**
         * Habilita a opção para exportação dos dados.
         */
        _this.exportButtons = false;
        /**
         * Recebe valores "true" ou "false" para habilitar ou desabilitar o botão "Carregar Mais Resultados"
         */
        _this.showMoreDisabled = 'false';
        /**
         * Habilita o botão "Remover" permitindo que o usuário possa remover uma linha do grid.
         */
        _this.showRemoveButton = false;
        /**
         * Habilita o botão para adicionar linhas
         */
        _this.addButton = false;
        /**
         * Habilita um botão "Carregar mais resultados" com evento de clique que chama um método do componente.
         */
        _this.showMore = new core.EventEmitter();
        /**
         * Evento de seleção de linha que chama um método do componente. Este atributo é utilizado em conjunto com o atributo "t-selectable".
         */
        _this.selectionChange = new core.EventEmitter();
        /**
         * Evento disparado ao salvar dados do modo de edição inline, contém o modelo que foi alterado.
         */
        _this.saveValue = new core.EventEmitter();
        _this.editedRowIndex = -1;
        _this.groups = [];
        _this.state = { skip: 0 };
        _this.sort = [];
        _this.addButtonCalled = false;
        _this.isCancelling = false;
        _this._parentRef = viewRef['_view']['component'];
        _this.allData = _this.allData.bind(_this);
        return _this;
    }
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.renderer.listen('document', 'click', function (_a) {
            var target = _a.target;
            _this.validateSaveEventInDocument(target);
        });
        if (!this.data) {
            this.data = [];
        }
        if (!this.columns) {
            this.columns = [];
        }
        else {
            this.defineColumnType();
        }
        if (this.sortable) {
            this.sortableObject = {
                allowUnsort: this.sortable,
                mode: 'single'
            };
        }
        if (this.groupable) {
            var /** @type {?} */ arraySize = this.columns.length;
            for (var /** @type {?} */ count = 0; count < arraySize; count++) {
                var /** @type {?} */ columnTemp = this.columns[count];
                if (this.groups.length < 2 && columnTemp.groupHeader) {
                    this.groups.push({ field: columnTemp.column });
                }
            }
            this.sortableObject = null;
            this.loadDataGroupable();
        }
        else {
            this.loadData();
        }
        if (!this.editable) {
            this.grid = null;
        }
    };
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.defineColumnType = function () {
        var /** @type {?} */ lookupTableType = {
            number: function (column) {
                column.type = 'numeric';
                column.format = undefined;
            },
            currency: function (column) {
                column.type = 'numeric';
                column.format = '{0:c}';
            },
            date: function (column) {
                column.type = 'date';
                if (column.format && column.format.trim().length > 0) {
                    column.format = "{0:" + column.format + "}";
                }
                else {
                    column.format = '{0:dd/MM/yyyy}';
                }
            },
            string: function (column) {
                column.type = 'text';
                column.format = undefined;
            }
        };
        this.columns.forEach(function (column) {
            if (column.type && lookupTableType.hasOwnProperty(column.type.trim().toLowerCase())) {
                lookupTableType[column.type.trim().toLowerCase()](column);
            }
            else {
                column.type = 'text';
            }
        });
    };
    /**
     * @param {?} el
     * @param {?} className
     * @return {?}
     */
    ThfGridComponent.prototype.isChildOf = function (el, className) {
        while (el && el.parentElement) {
            if (this.hasClass(el.parentElement, className)) {
                return true;
            }
            el = el.parentElement;
        }
        return false;
    };
    /**
     * @param {?} el
     * @param {?} className
     * @return {?}
     */
    ThfGridComponent.prototype.hasClass = function (el, className) {
        return new RegExp(className).test(el.className);
    };
    /**
     * @param {?} target
     * @return {?}
     */
    ThfGridComponent.prototype.validateSaveEventInDocument = function (target) {
        if (!this.isChildOf(target, 'k-grid-content') && !this.isChildOf(target, 'k-grid-toolbar')) {
            this.saveClick();
        }
    };
    /**
     * @param {?} sort
     * @return {?}
     */
    ThfGridComponent.prototype.sortChange = function (sort) {
        this.sort = sort;
        this.loadData();
    };
    /**
     * @param {?} state
     * @return {?}
     */
    ThfGridComponent.prototype.dataStateChange = function (state) {
        // Esta condição foi realizada para não deixar adicionar mais de 2 grupos devido
        // a um problema no kendo grid. Ja tem um chamado aberto para este problema.
        if (state.group.length > 2) {
            state.group.splice(0, 1);
        }
        this.state = state;
        this.gridView = kendoDataQuery.process(this.data, this.state);
    };
    /**
     * @param {?} dataUpdated
     * @return {?}
     */
    ThfGridComponent.prototype.updateIndex = function (dataUpdated) {
        // A variavel "this.data" é a fonte de dados principal,
        // na linha abaixo eu estou atualizando a fonte de dados
        // principal com a fonte de dados ordenada para os indices
        // não se perderem na hora de salvar uma edição.
        for (var /** @type {?} */ i = 0, /** @type {?} */ dataLength = dataUpdated.length; i < dataLength; i++) {
            this.data[i] = dataUpdated[i];
        }
    };
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.loadData = function () {
        // A variavel "this.data" é a fonte de dados principal,
        // na linha abaixo eu estou atualizando a fonte de dados
        // principal com a fonte de dados ordenada para os indices
        // não se perderem na hora de salvar uma edição.
        this.updateIndex(kendoDataQuery.orderBy(this.data, this.sort));
        this.gridView = {
            data: this.data,
            total: this.data.length
        };
    };
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.createFormGroup = function () {
        var /** @type {?} */ group = {};
        for (var /** @type {?} */ count = 0; count < this.columns.length; count++) {
            var /** @type {?} */ columnTemp = this.columns[count];
            if (!columnTemp.checkbox) {
                if (columnTemp.required) {
                    group[columnTemp.column] = new forms.FormControl('', forms.Validators.required);
                }
                else {
                    group[columnTemp.column] = new forms.FormControl('');
                }
            }
        }
        this.formGroup = new forms.FormGroup(group);
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    ThfGridComponent.prototype.addHandler = function (_a) {
        var sender = _a.sender;
        if (this.editedRowIndex >= 0) {
            this.closeEditor(sender, this.editedRowIndex);
        }
        this.addButtonCalled = true;
        this.createFormGroup();
        if (this.addAction) {
            if (this.executeFunctionValidation(this.addAction, this.formGroup.value)) {
                this.formGroup.setValue(this.formGroup.value);
            }
            else {
                return;
            }
        }
        sender.addRow(this.formGroup);
    };
    /**
     * @param {?} func
     * @param {?} param
     * @return {?}
     */
    ThfGridComponent.prototype.executeFunctionValidation = function (func, param) {
        return (func && this._parentRef[func](param));
    };
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.saveLine = function () {
        if (this.editable) {
            var /** @type {?} */ newRowValue = void 0;
            if (this.editedRowIndex >= 0) {
                newRowValue = Object.assign(this.data[this.editedRowIndex], this.formGroup.value);
                this.data[this.editedRowIndex] = newRowValue;
            }
            else if (this.addButtonCalled && this.formGroup) {
                newRowValue = this.formGroup.value;
                this.data.push(newRowValue);
                this.addButtonCalled = false;
            }
            if (this.isGroup()) {
                this.loadDataGroupable();
            }
            this.saveValue.emit({ data: newRowValue });
        }
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    ThfGridComponent.prototype.editClickHandler = function (_a) {
        var rowIndex = _a.rowIndex, dataItem = _a.dataItem;
        if (this.isValidForm()) {
            return;
        }
        if (this.executeFunctionValidation(this.saveAction, { data: this.data[this.editedRowIndex] })) {
            this.saveLine();
        }
        // Verifica se está utilizando agrupamento e busca o indice atualizado
        // do objeto que está sendo editado no momento.
        if (this.isGroup()) {
            rowIndex = this.getRowIndex(this.data, dataItem);
        }
        this.editHandler({
            sender: this.grid,
            rowIndex: rowIndex,
            dataItem: dataItem
        });
    };
    /**
     * @param {?} arr
     * @param {?} searchFor
     * @return {?}
     */
    ThfGridComponent.prototype.getRowIndex = function (arr, searchFor) {
        var /** @type {?} */ keys = Object.keys(searchFor);
        for (var /** @type {?} */ i = 0; i < arr.length; i++) {
            var /** @type {?} */ isEqual = true;
            for (var _i = 0, keys_1 = keys; _i < keys_1.length; _i++) {
                var key = keys_1[_i];
                if (arr[i][key] !== searchFor[key]) {
                    isEqual = false;
                    break;
                }
            }
            if (isEqual) {
                return i;
            }
        }
        return -1;
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    ThfGridComponent.prototype.editHandler = function (_a) {
        var sender = _a.sender, rowIndex = _a.rowIndex, dataItem = _a.dataItem;
        if (!this.editable) {
            return;
        }
        this.closeEditor(sender);
        this.editedProducted = Object.assign({}, dataItem);
        this.formGroup = new forms.FormGroup({});
        for (var _i = 0, _b = Object.keys(dataItem); _i < _b.length; _i++) {
            var key = _b[_i];
            var /** @type {?} */ control = void 0;
            var /** @type {?} */ columnTemp = this.getColumn(key);
            if (columnTemp && columnTemp.editable) {
                if (columnTemp.required) {
                    control = new forms.FormControl(dataItem[key], forms.Validators.required);
                }
                else {
                    control = new forms.FormControl(dataItem[key]);
                }
                this.formGroup.addControl(key, control);
            }
        }
        this.editedRowIndex = rowIndex;
        sender.editRow(rowIndex, this.formGroup);
    };
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.saveClick = function () {
        if (this.addButtonCalled && this.saveAction && this.formGroup) {
            if (!this.executeFunctionValidation(this.saveAction, { data: this.formGroup.value })) {
                return;
            }
        }
        else if (this.editedRowIndex >= 0 && this.saveAction) {
            if (!this.executeFunctionValidation(this.saveAction, { data: this.data[this.editedRowIndex] })) {
                this.closeEditor(this.grid, this.editedRowIndex);
                return;
            }
        }
        if (this.isValidForm()) {
            return;
        }
        this.saveLine();
        this.closeEditor(this.grid);
        if (this.isGroup()) {
            this.loadDataGroupable();
        }
        else {
            this.loadData();
        }
    };
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.isGroup = function () {
        return (this.groups && this.groups.length > 0);
    };
    /**
     * @param {?} key
     * @return {?}
     */
    ThfGridComponent.prototype.getColumn = function (key) {
        for (var /** @type {?} */ index = 0; index < this.columns.length; index++) {
            var /** @type {?} */ element = this.columns[index];
            if (element.column === key) {
                return element;
            }
        }
    };
    /**
     * @param {?} grid
     * @param {?=} rowIndex
     * @return {?}
     */
    ThfGridComponent.prototype.closeEditor = function (grid, rowIndex) {
        rowIndex = this.editedRowIndex;
        if (grid) {
            grid.closeRow(rowIndex);
        }
        this.editedRowIndex = undefined;
        this.formGroup = undefined;
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    ThfGridComponent.prototype.saveHandler = function (_a) {
        var sender = _a.sender, rowIndex = _a.rowIndex, formGroup = _a.formGroup, isNew = _a.isNew;
        var /** @type {?} */ item = formGroup.value;
        if (isNew && !this.editable) {
            this.data.push(item);
        }
        else {
            this.data[rowIndex] = Object.assign(this.data[rowIndex], item);
        }
        if (this.isGroup()) {
            this.loadDataGroupable();
        }
        else {
            this.loadData();
        }
        sender.closeRow(rowIndex);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ThfGridComponent.prototype.cancelPropagation = function (event) {
        event.stopPropagation();
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    ThfGridComponent.prototype.cancelHandler = function (_a) {
        var sender = _a.sender, rowIndex = _a.rowIndex;
        if (this.editedProducted != null) {
            this.data[rowIndex] = this.editedProducted;
            this.editedProducted = null;
        }
        if (this.isGroup()) {
            this.loadDataGroupable();
        }
        else {
            this.loadData();
        }
        this.closeEditor(sender);
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    ThfGridComponent.prototype.removeHandler = function (_a) {
        var rowIndex = _a.rowIndex;
        if (this.removeAction) {
            if (!this.executeFunctionValidation(this.removeAction, { data: this.data[rowIndex] })) {
                return;
            }
        }
        this.data.splice(rowIndex, 1);
        if (this.groupable) {
            this.loadDataGroupable();
        }
        else {
            this.loadData();
        }
    };
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.isValidForm = function () {
        return (this.formGroup && !this.formGroup.valid);
    };
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.allData = function () {
        var /** @type {?} */ result = {
            data: kendoDataQuery.process(this.data, {}).data
        };
        return result;
    };
    /**
     * @param {?} event
     * @param {?} index
     * @param {?} data
     * @param {?} column
     * @return {?}
     */
    ThfGridComponent.prototype.changeValueCheckbox = function (event, index, data, column) {
        if (!this.editable) {
            event.target.checked = !event.target.checked;
            return;
        }
        data[column] = event.target.checked;
        this.data[index] = Object.assign(data);
        this.saveValue.emit({ data: this.data[index] });
    };
    /**
     * @param {?} e
     * @return {?}
     */
    ThfGridComponent.prototype.onSelectionChange = function (e) {
        this.selectionChange.emit({ data: this.gridView.data[e.index] });
    };
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.onShowMore = function () {
        this.showMore.emit(null);
        if (this.groupable) {
            this.loadDataGroupable();
        }
        else {
            this.loadData();
        }
    };
    /**
     * @param {?} groups
     * @return {?}
     */
    ThfGridComponent.prototype.groupChange = function (groups) {
        this.groups = groups;
        this.loadDataGroupable();
    };
    /**
     * @return {?}
     */
    ThfGridComponent.prototype.loadDataGroupable = function () {
        this.gridView = kendoDataQuery.process(this.data, { group: this.groups });
        this.dataArrayOrdered = [];
        this.getObjects(this.gridView.data);
        this.updateIndex(this.dataArrayOrdered);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ThfGridComponent.prototype.getObjects = function (data) {
        var _this = this;
        data.forEach(function (value) {
            if (value.items instanceof Array) {
                _this.getObjects(value.items);
            }
            else {
                _this.dataArrayOrdered.push(value);
            }
        });
    };
    ThfGridComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'thf-grid',
                    encapsulation: core.ViewEncapsulation.None, template: "<kendo-grid  [data]=\"gridView\"  [sortable]=\"sortableObject\" [sort]=\"sort\" [groupable]=\"groupable\" [group]=\"groups\" [selectable]=\"selectable\" (dataStateChange)=\"dataStateChange($event)\" (groupChange)=\"groupChange($event)\" (selectionChange)=\"onSelectionChange($event)\" (sortChange)=\"sortChange($event)\" (add)=\"addHandler($event)\" (cancel)=\"cancelHandler($event)\" (cellClick)=\"editClickHandler($event)\" (edit)=\"editHandler($event)\" (remove)=\"removeHandler($event)\" (save)=\"saveHandler($event)\" style=\"height: 70%\"> <kendo-grid-messages  i18n-noRecords=\"\" noRecords=\"Não há registros\" i18n-groupPanelEmpty=\"\" groupPanelEmpty=\"Arraste a coluna até o cabeçalho e solte para agrupar por esta coluna\"> </kendo-grid-messages> <ng-template kendoGridToolbarTemplate *ngIf=\"addButton || exportButtons\" > <button kendoGridAddCommand *ngIf=\"addButton\" class=\"k-primary\">Adicionar</button> <button type=\"button\" kendoGridExcelCommand *ngIf=\"exportButtons\" ><span class=\"k-icon k-i-file-excel\"></span></button> <button kendoGridPDFCommand *ngIf=\"exportButtons\"><span class='k-icon k-i-file-pdf'></span></button> </ng-template> <kendo-grid-column *ngFor=\"let col of columns; let i = index\" [field]=\"col.column\" [title]=\"col.label\" [width]=\"col.width\" [filter]=\"col.filter\" [format]=\"col.format\" [editor]=\"col.type\"> <ng-template kendoGridCellTemplate let-data *ngIf=\"col.checkbox\" let-rowIndex=\"rowIndex\"> <input type=\"checkbox\" id=\"chkbox_{{rowIndex}}\" name=\"chkbox_{{col.column}}_{{rowIndex}}\" [checked]=\"data[col.column]\" (change)=\"changeValueCheckbox($event, rowIndex, data, col.column)\" class=\"thf-grid thf-grid-checkbox\" /> <label for=\"chkbox_{{rowIndex}}\"></label> </ng-template> <ng-template kendoGridGroupHeaderTemplate let-value=\"value\" *ngIf=\"col.groupHeader\"> {{value}} </ng-template> </kendo-grid-column> <kendo-grid-command-column title=\"\" width=\"80\"> <ng-template kendoGridCellTemplate let-isNew=\"isNew\"> <button kendoGridRemoveCommand class=\"k-primary\" *ngIf=\"showRemoveButton\" >Remover</button> <button kendoGridCancelCommand class=\"k-primary\" (click)=\"cancelPropagation($event)\">{{ isNew ? 'Descartar' : 'Cancelar' }} </button> </ng-template> </kendo-grid-command-column> <kendo-grid-excel fileName=\"spreadsheet.xlsx\" [fetchData]=\"allData\"></kendo-grid-excel> <kendo-grid-pdf fileName=\"grid.pdf\" [allPages]=\"true\"> <kendo-grid-pdf-margin top=\"1cm\" left=\"1cm\" right=\"1cm\" bottom=\"1cm\"></kendo-grid-pdf-margin> </kendo-grid-pdf> </kendo-grid> <div class=\"thf-grid-show-more\" *ngIf=\"showMore.observers.length > 0\"> <div style=\"text-align: center;\"> <thf-button t-label=\"Carregar mais resultados\" (t-click)=\"onShowMore()\" t-id=\"btnShowMoreId\" [t-disabled]=\"showMoreDisabled\"></thf-button> </div> </div> ", styles: [".k-grid .k-state-selected { background-color: skyblue !important; color: #000000; } .k-grid .k-alt.k-state-selected { background-color: skyblue !important; color: #000000; } .thf-grid.thf-grid-checkbox { width: 20px; height: 20px; border-radius: 2px; box-shadow: inset 0 1px 8px 0 rgba(0, 0, 0, 0.1); border: solid 1px rgba(255, 255, 255, 0.6); } .thf-grid.thf-grid-checkbox:checked { width: 20px; height: 20px; border-radius: 2px; } input[type=checkbox] { visibility: hidden; position: absolute; } input[type=checkbox] + label:before { width: 20px; height: 20px; margin-top: -2px; margin-right: 8px; display: inline-block; vertical-align: middle; background-color: #ffffff; box-shadow: inset 0px 2px 8px 0 #edefef; border: solid 1px #b7bebf; content: ' '; } input[type=checkbox]:checked + label:before { background-color: #0c9abe; box-shadow: none; color: white; font-family: 'TotvsIcon' !important; line-height: 18px; content: '\\e937'; padding-left: 1px; } input[type=checkbox]:disabled + label:before { background-color: #edefef; border: solid 1px #b7bebf; } input[type=checkbox]:checked:disabled + label:before { color: #b7bebf; border: solid 1px #b6bdbf; } input[type=checkbox] + label:before { border-radius: 2px; } .thf-grid-show-more { width: 100%; margin-top: 10px; margin-bottom: 10px; } "]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfGridComponent.ctorParameters = function () { return [
        { type: core.ViewContainerRef, },
        { type: core.Renderer2, },
    ]; };
    ThfGridComponent.propDecorators = {
        'columns': [{ type: core.Input, args: ['t-columns',] },],
        'sortable': [{ type: core.Input, args: ['t-sortable',] },],
        'selectable': [{ type: core.Input, args: ['t-selectable',] },],
        'editable': [{ type: core.Input, args: ['t-editable',] },],
        'groupable': [{ type: core.Input, args: ['t-groupable',] },],
        'exportButtons': [{ type: core.Input, args: ['t-show-export-buttons',] },],
        'showMoreDisabled': [{ type: core.Input, args: ['t-show-more-disabled',] },],
        'showRemoveButton': [{ type: core.Input, args: ['t-show-remove-button',] },],
        'addButton': [{ type: core.Input, args: ['t-show-add-button',] },],
        'saveAction': [{ type: core.Input, args: ['t-save-action',] },],
        'removeAction': [{ type: core.Input, args: ['t-remove-action',] },],
        'addAction': [{ type: core.Input, args: ['t-add-action',] },],
        'showMore': [{ type: core.Output, args: ['t-show-more',] },],
        'selectionChange': [{ type: core.Output, args: ['t-selection-change',] },],
        'saveValue': [{ type: core.Output, args: ['t-save-value',] },],
        'grid': [{ type: core.ViewChild, args: [kendoAngularGrid.GridComponent,] },],
    };
    return ThfGridComponent;
}(thfGridBase_component.ThfGridBaseComponent));

var __extends$2 = (undefined && undefined.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 *
 * \@example
 * <example name='thf-modal' title="Totvs Modal Labs">
 *  <file name='sample-thf-modal.component.html'>
 *    <thf-button
 *      t-label="Chamar Modal"
 *      (t-click)="openPedidoModal()"
 *      t-primary="true">
 *    </thf-button>
 *
 *    <form #f="ngForm">
 *
 *      <thf-input
 *        t-label="Titulo"
 *        t-help="Informe o título da modal"
 *        t-placeholder="Título"
 *        name="title"
 *        [(ngModel)]="modalTitle"
 *        t-maxlength="30"
 *        t-clean>
 *      </thf-input>
 *
 *      <thf-input
 *        t-label="Texto"
 *        t-help="Informe o texto da modal"
 *        t-placeholder="Texto"
 *        name="text"
 *        [(ngModel)]="modalText"
 *        t-maxlength="100"
 *        t-focus
 *        t-clean>
 *      </thf-input>
 *
 *      <thf-radio-group
 *        name="modalSize"
 *        t-required
 *        t-label="Tamanho da modal"
 *        [t-options]="modalSizeOptions"
 *        [(ngModel)]="modalSize">
 *      </thf-radio-group>
 *
 *      <thf-checkbox-group
 *        name="configurationCheckbox"
 *        t-label="Configurações"
 *        t-help="Permite fazer ajustes na modal"
 *        [t-options]="configurationOptions"
 *        [(ngModel)]="configurationModel">
 *      </thf-checkbox-group>
 *
 *    </form>
 *
 *    <thf-modal
 *      [t-title]="modalTitle"
 *      [t-primary-action]="primaryAction"
 *      [t-secondary-action]="secondaryAction"
 *      [t-size]="modalSize"
 *      [t-click-out]="configurationModel.includes('click-out')"
 *      [t-hide-close]="configurationModel.includes('hide-close')">
 *      {{modalText}}
 *    </thf-modal>
 *  </file>
 *  <file name='sample-thf-modal.component.ts'>
 *    import { Component, ViewChild } from '\@angular/core';
 *
 *    import { ThfModalAction } from "\@totvs/thf-core/components/thf-modal-base/thf-modal-action.interface";
 *    import { ThfModalComponent } from '\@totvs/thf-web/components/thf-modal/thf-modal.component';
 *    import { ThfRadioGroupOption, ThfCheckboxGroupOption } from '\@totvs/thf-core/components/thf-field-base';
 *
 *    \@Component({
 *      selector: 'sample-thf-modal',
 *      templateUrl: './sample-thf-modal.component.html'
 *    })
 *    export class SampleThfModalComponent {
 *
 *      \@ViewChild(ThfModalComponent) thfModal: ThfModalComponent;
 *
 *      primaryAction: ThfModalAction = {
 *        action: () => {
 *          this.thfModal.close();
 *        },
 *        label: 'Confirmar'
 *      };
 *
 *      secondaryAction: ThfModalAction = {
 *        action: () => {
 *           this.thfModal.close();
 *        },
 *        label: 'Cancelar'
 *      };
 *
 *      modalSize = 'sm';
 *      modalText = 'Somos TOTVERS!!!';
 *      modalTitle = 'Totvs';
 *
 *      configurationOptions: Array<ThfCheckboxGroupOption> = [
 *        {value: 'click-out', label: 'Permitir o fechamento ao clicar fora da modal' },
 *        {value: 'hide-close', label: 'Esconder o ícone de fechar a modal' }
 *      ];
 *      configurationModel: Array<any> = [];
 *
 *      modalSizeOptions: Array<ThfRadioGroupOption> = [
 *        { label: 'Pequeno', value: 'sm' },
 *        { label: 'Médio', value: 'md' },
 *        { label: 'Grande', value: 'lg' },
 *        { label: 'Extra grande', value: 'xl' },
 *        { label: 'Automático', value: 'auto' },
 *      ];
 *
 *      openPedidoModal() {
 *        this.thfModal.open();
 *      }
 *
 *    }
 *  </file>
 * </example>
 */
var ThfModalComponent = (function (_super) {
    __extends$2(ThfModalComponent, _super);
    function ThfModalComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.focusableElements = 'input, select, textarea, button, a';
        return _this;
    }
    /**
     * @return {?}
     */
    ThfModalComponent.prototype.open = function () {
        _super.prototype.open.call(this);
        this.handleFocus();
    };
    /**
     * @return {?}
     */
    ThfModalComponent.prototype.close = function () {
        _super.prototype.close.call(this);
        document.removeEventListener('focus', this.focusFunction, true);
    };
    /**
     * @return {?}
     */
    ThfModalComponent.prototype.handleFocus = function () {
        var _this = this;
        setTimeout(function () {
            if (_this.modalContent) {
                _this.initFocus();
                document.addEventListener('focus', _this.focusFunction, true);
            }
        });
    };
    /**
     * @return {?}
     */
    ThfModalComponent.prototype.initFocus = function () {
        var _this = this;
        this.focusFunction = function (event) {
            var /** @type {?} */ modalElement = _this.modalContent.nativeElement;
            if (!modalElement.contains(event.target)) {
                event.stopPropagation();
                _this.firstElement.focus();
            }
        };
        this.setFirstElement();
        if (!this.hideClose) {
            var /** @type {?} */ firstFieldElement = this.modalContent.nativeElement.querySelectorAll(this.focusableElements)[1];
            firstFieldElement.focus();
        }
        else {
            this.firstElement.focus();
        }
    };
    /**
     * @return {?}
     */
    ThfModalComponent.prototype.setFirstElement = function () {
        this.firstElement = this.modalContent.nativeElement.querySelector(this.focusableElements);
    };
    ThfModalComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'thf-modal', template: "<div class=\"thf-modal\" *ngIf=\"!isHidden\"> <div class=\"thf-modal-overlay\" (click)=\"onClickOut()\"></div> <div #modalContent class=\"thf-modal-content thf-modal-{{ modalSize }}\" tabindex=\"-1\"> <div class=\"thf-modal-header\"> <div class=\"thf-modal-title\"> <a class=\"thf-modal-header-close-button\" *ngIf=\"!hideClose\" (click)=\"close()\" tabindex=\"0\" > <span class=\"thf-icon thf-icon-close\"></span> </a> {{ title }} </div> </div> <div class=\"thf-modal-body\"> <ng-content></ng-content> </div> <div class=\"thf-modal-footer\"> <thf-button *ngIf=\"secondaryAction\" class=\"thf-modal-footer-button\" [t-label]=\"secondaryAction.label\" (t-click)=\"secondaryAction.action()\" t-primary=\"false\"> </thf-button> <thf-button *ngIf=\"primaryAction\" class=\"thf-modal-footer-button\" [t-label]=\"primaryAction.label\" (t-click)=\"primaryAction.action()\" t-primary=\"true\"> </thf-button> </div> </div> </div> ", styles: [".thf-modal { font-family: NunitoSans; position: relative; z-index: 50; } .thf-modal-overlay:before { content: ''; position: fixed; background-color: #052e3e; top: 0; bottom: 0; right: 0; left: 0; opacity: 0.7; } .thf-modal-sm { width: 50% !important; } .thf-modal-md { width: 70% !important; } .thf-modal-lg { width: 80% !important; } .thf-modal-xl { width: 90% !important; } .thf-modal-auto { width: auto !important; } @media screen and (max-width: 480px) { .thf-modal-content { position: fixed; left: 50%; top: 50%; transform: translate(-50%, -50%); background-color: #ffffff; box-shadow: 0px 4px 16px 0 rgba(7, 9, 9, 0.24); max-height: 90%; width: 90%; } } @media screen and (min-width: 481px) and (max-width: 960px) { .thf-modal-content { position: fixed; left: 50%; top: 50%; transform: translate(-50%, -50%); background-color: #ffffff; box-shadow: 0px 4px 16px 0 rgba(7, 9, 9, 0.24); max-height: 90%; width: 80%; } } @media screen and (min-width: 961px) { .thf-modal-content { position: fixed; left: 50%; top: 50%; transform: translate(-50%, -50%); background-color: #ffffff; box-shadow: 0px 4px 16px 0 rgba(7, 9, 9, 0.24); max-height: 90%; width: 70%; } } .thf-modal-header { border-radius: 3px; padding-top: 18px; padding-left: 24px; padding-right: 24px; padding-bottom: 12px; border: 0px; } .thf-modal-title { font-size: 34px; font-weight: 200; line-height: 27px; text-align: left; color: #4a5c60; } .thf-modal-header-close-button { float: right; border-radius: 1px; text-align: right; color: #0c9abe; cursor: pointer; text-decoration: none; } .thf-modal-body { margin: 12px 24px 12px 24px; font-size: 14px; line-height: 1.71; text-align: left; color: #425357; max-height: calc(100vh - 210px); overflow-y: auto; } .thf-modal-footer { font-family: NunitoSans; background-color: #ffffff; border-radius: 3px; border: 0px; text-align: right; padding-top: 12px; padding-left: 0px; padding-bottom: 24px; padding-right: 24px; } .thf-modal-footer-button { margin-left: 16px; margin-top: 5px; } ::-webkit-scrollbar { width: 4px; height: 50px; opacity: 0.5; border-radius: 2px; } ::-webkit-scrollbar-thumb { background-color: #0c9abe; } "]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfModalComponent.ctorParameters = function () { return []; };
    ThfModalComponent.propDecorators = {
        'modalContent': [{ type: core.ViewChild, args: ['modalContent', { read: core.ElementRef },] },],
    };
    return ThfModalComponent;
}(thfModalBase.ThfModalBaseComponent));

/**
 * \@description
 * Módulo do componente thf-modal
 */
var ThfModalModule = (function () {
    function ThfModalModule() {
    }
    ThfModalModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule,
                        ThfButtonModule
                    ],
                    declarations: [
                        ThfModalComponent
                    ],
                    exports: [ThfModalComponent],
                    providers: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfModalModule.ctorParameters = function () { return []; };
    return ThfModalModule;
}());

var hasClass = function (el, className) { return new RegExp(className).test(el.className); };
var isChildOf = function (el, className) {
    while (el && el.parentElement) {
        if (hasClass(el.parentElement, className)) {
            return true;
        }
        el = el.parentElement;
    }
    return false;
};
var eq = function (s1, s2) { return s1.toLowerCase() === s2.toLowerCase(); };
var closest = function (el, nodeName) {
    while (el && el.parentElement) {
        if (eq(el.nodeName, nodeName) || eq(el.parentElement.nodeName, nodeName)) {
            return el.parentElement;
        }
        el = el.parentElement;
    }
    return null;
};
var ThfRowClickDirective = (function () {
    /**
     * @param {?} el
     * @param {?} renderer
     */
    function ThfRowClickDirective(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.editRow = new core.EventEmitter();
        this.saveRow = new core.EventEmitter();
    }
    /**
     * @return {?}
     */
    ThfRowClickDirective.prototype.ngOnInit = function () {
        var _this = this;
        this.renderer.listen(this.el.nativeElement, 'click', function (_a) {
            var target = _a.target;
            var /** @type {?} */ tr = closest(target, 'tr');
            if (tr && !hasClass(tr, 'k-grid-edit-row') && isChildOf(target, 'k-grid-content')) {
                _this.editRow.emit(tr.rowIndex);
            }
        });
        this.renderer.listen('document', 'click', function (_a) {
            var target = _a.target;
            if (!isChildOf(target, 'k-grid-content') && !isChildOf(target, 'k-grid-toolbar')) {
                _this.saveRow.emit();
            }
        });
    };
    ThfRowClickDirective.decorators = [
        { type: core.Directive, args: [{ selector: '[row-click]' },] },
    ];
    /**
     * @nocollapse
     */
    ThfRowClickDirective.ctorParameters = function () { return [
        { type: core.ElementRef, },
        { type: core.Renderer2, },
    ]; };
    ThfRowClickDirective.propDecorators = {
        'editRow': [{ type: core.Output },],
        'saveRow': [{ type: core.Output },],
    };
    return ThfRowClickDirective;
}());

/**
 * \@description
 * Módulo do componente thf-grid
 */
var ThfGridModule = (function () {
    function ThfGridModule() {
    }
    ThfGridModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule,
                        kendoAngularGrid.GridModule,
                        kendoAngularGrid.ExcelModule,
                        kendoAngularGrid.PDFModule,
                        ThfButtonModule,
                        ThfModalModule
                    ],
                    declarations: [
                        ThfGridComponent,
                        ThfRowClickDirective
                    ],
                    exports: [ThfGridComponent],
                    providers: [],
                    schemas: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfGridModule.ctorParameters = function () { return []; };
    return ThfGridModule;
}());

exports.ThfGridModule = ThfGridModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
