// @totvs/thf-web/components/thf-field/thf-field-container/thf-field-container.component.js
/**
 * Componente de uso interno, responsável por atribuir uma label para o campo
 */
import { Component, Input } from '@angular/core';
var ThfFieldContainerComponent = (function () {
    function ThfFieldContainerComponent() {
        this.optional = {
            pt: '(Opcional)',
            en: '(Optional)',
            es: '(Opcional)'
        };
    }
    /**
     * @return {?}
     */
    ThfFieldContainerComponent.prototype.getOptional = function () {
        var /** @type {?} */ lang = (navigator.language) ? navigator.language.toLowerCase().slice(0, 2) : 'pt';
        return (this.optional[lang]) ? this.optional[lang] : '';
    };
    /**
     * @return {?}
     */
    ThfFieldContainerComponent.prototype.ngOnInit = function () {
    };
    ThfFieldContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-field-container', template: " <div style=\"display:block\"> <span class=\"thf-label\">{{label}}</span> <span *ngIf=\"opcional\" class=\"thf-opcional\">{{getOptional()}}</span> <br> <span *ngIf=\"help!==''\" class=\"thf-label-help\">{{help}}</span> <ng-content></ng-content> </div> ", styles: [".thf-label { font-size: 14px; line-height: 2; text-align: left; color: #4a5c60; font-family: 'NunitoSans-Bold'; } .thf-opcional { font-size: 12px; line-height: 2; color: #4a5c60; font-family: NunitoSans; float: right !important; text-align: bottom; } .thf-label-help { font-size: 12px; line-height: 1.67; text-align: left; color: #4a5c60; font-family: NunitoSans; } "]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfFieldContainerComponent.ctorParameters = function () { return []; };
    ThfFieldContainerComponent.propDecorators = {
        'label': [{ type: Input, args: ['t-label',] },],
        'help': [{ type: Input, args: ['t-help',] },],
        'opcional': [{ type: Input, args: ['t-opcional',] },],
    };
    return ThfFieldContainerComponent;
}());
export { ThfFieldContainerComponent };
function ThfFieldContainerComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfFieldContainerComponent.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfFieldContainerComponent.ctorParameters;
    /** @type {?} */
    ThfFieldContainerComponent.propDecorators;
    /**
     * Label do campo
     * @type {?}
     */
    ThfFieldContainerComponent.prototype.label;
    /**
     * Texto de apoio do campo
     * @type {?}
     */
    ThfFieldContainerComponent.prototype.help;
    /**
     * Label do campo
     * @type {?}
     */
    ThfFieldContainerComponent.prototype.opcional;
    /** @type {?} */
    ThfFieldContainerComponent.prototype.optional;
}
