// @totvs/thf-web/interceptors/interceptors.module.js
import { NgModule } from '@angular/core';
import { ThfHttpInterceptorModule } from './thf-http-interceptor/thf-http-interceptor.module';
var ThfInterceptorsModule = (function () {
    function ThfInterceptorsModule() {
    }
    ThfInterceptorsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        ThfHttpInterceptorModule
                    ],
                    exports: [
                        ThfHttpInterceptorModule,
                    ],
                    providers: [],
                    bootstrap: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfInterceptorsModule.ctorParameters = function () { return []; };
    return ThfInterceptorsModule;
}());
export { ThfInterceptorsModule };
function ThfInterceptorsModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfInterceptorsModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfInterceptorsModule.ctorParameters;
}
