// @totvs/thf-web/components/components.module.js
import { NgModule } from '@angular/core';
import { ThfAvatarModule } from './thf-avatar/thf-avatar.module';
import { ThfButtonModule } from './thf-button/thf-button.module';
import { ThfChartModule } from './thf-chart/thf-chart.module';
import { ThfDropdownModule } from './thf-dropdown/thf-dropdown.module';
import { ThfFieldModule } from './thf-field/thf-field.module';
import { ThfGridModule } from './thf-grid/thf-grid.module';
import { ThfInfoModule } from './thf-info/thf-info.module';
import { ThfListModule } from './thf-list/thf-list.module';
import { ThfMenuModule } from './thf-menu/thf-menu.module';
import { ThfModalModule } from './thf-modal/thf-modal.module';
import { ThfPageModule } from './thf-page/thf-page.module';
import { ThfTableModule } from './thf-table/thf-table.module';
import { ThfTabsModule } from './thf-tabs/thf-tabs.module';
import { ThfToolbarModule } from './thf-toolbar/thf-toolbar.module';
import { ThfWidgetModule } from './thf-widget/thf-widget.module';
var ThfComponentsModule = (function () {
    function ThfComponentsModule() {
    }
    ThfComponentsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        ThfAvatarModule,
                        ThfButtonModule,
                        ThfChartModule,
                        ThfDropdownModule,
                        ThfFieldModule,
                        ThfGridModule,
                        ThfInfoModule,
                        ThfListModule,
                        ThfMenuModule,
                        ThfModalModule,
                        ThfPageModule,
                        ThfTableModule,
                        ThfTabsModule,
                        ThfToolbarModule,
                        ThfWidgetModule
                    ],
                    exports: [
                        ThfAvatarModule,
                        ThfButtonModule,
                        ThfChartModule,
                        ThfDropdownModule,
                        ThfFieldModule,
                        ThfGridModule,
                        ThfInfoModule,
                        ThfListModule,
                        ThfMenuModule,
                        ThfModalModule,
                        ThfPageModule,
                        ThfTabsModule,
                        ThfTableModule,
                        ThfToolbarModule,
                        ThfWidgetModule
                    ],
                    providers: [],
                    bootstrap: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfComponentsModule.ctorParameters = function () { return []; };
    return ThfComponentsModule;
}());
export { ThfComponentsModule };
function ThfComponentsModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfComponentsModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfComponentsModule.ctorParameters;
}
