// @totvs/thf-web/components/thf-toolbar/thf-toolbar.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThfToolbarComponent } from './thf-toolbar.component';
import { ThfAvatarModule } from './../thf-avatar/thf-avatar.module';
/**
 * \@description
 * Módulo do componente thf-toolbar
 */
var ThfToolbarModule = (function () {
    function ThfToolbarModule() {
    }
    ThfToolbarModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        ThfAvatarModule
                    ],
                    declarations: [
                        ThfToolbarComponent
                    ],
                    exports: [ThfToolbarComponent],
                    providers: [],
                    schemas: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfToolbarModule.ctorParameters = function () { return []; };
    return ThfToolbarModule;
}());
export { ThfToolbarModule };
function ThfToolbarModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfToolbarModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfToolbarModule.ctorParameters;
}
