// @progress/kendo-angular-charts/dist/es/common/chart-instance-observer.js
import * as tslib_1 from "tslib";
import { InstanceObserver } from '@progress/kendo-charts';
/**
 * @hidden
 */
var ChartInstanceObserver = (function (_super) {
    tslib_1.__extends(ChartInstanceObserver, _super);
    function ChartInstanceObserver(ngZone, observer) {
        var _this = _super.call(this, observer) || this;
        _this.ngZone = ngZone;
        _this.handlerMap = {
            hideTooltip: 'onHideTooltip',
            legendItemClick: 'onLegendItemClick',
            render: 'onRender',
            showTooltip: 'onShowTooltip',
            init: 'onInit'
        };
        _this.outOfZone = ['onInit', 'trigger'];
        return _this;
    }
    ChartInstanceObserver.prototype.callObserver = function (fnName) {
        var _this = this;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var isDefaultPrevented = false;
        if (this.outOfZone.indexOf(fnName) >= 0) {
            isDefaultPrevented = this.observer[fnName].apply(this.observer, args);
        }
        else {
            this.ngZone.run(function () {
                isDefaultPrevented = _this.observer[fnName].apply(_this.observer, args);
            });
        }
        return isDefaultPrevented;
    };
    return ChartInstanceObserver;
}(InstanceObserver));
export { ChartInstanceObserver };
