// @totvs/thf-web/components/thf-page/thf-page-content/thf-page-content.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, Renderer } from '@angular/core';
import { ThfPageContentBaseComponent } from '@totvs/thf-core/components/thf-page-base';
/**
 * \@docsPrivate
 *
 */
var ThfPageContentComponent = (function (_super) {
    __extends(ThfPageContentComponent, _super);
    /**
     * @param {?} renderer
     */
    function ThfPageContentComponent(renderer) {
        var _this = _super.call(this) || this;
        _this.renderer = renderer;
        _this.height = '90%';
        _this.marginTop = '';
        _this.overflowY = 'none';
        renderer.listenGlobal('window', 'resize', function (event) {
            _this.setHeightContent(_this.pageHeaderElement, _this.bottomActions, _this.toolbarElement);
        });
        return _this;
    }
    /**
     * @return {?}
     */
    ThfPageContentComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        setTimeout(function () {
            _this.pageHeaderElement = (document.querySelector('div.thf-page-header'));
            _this.bottomActions = (document.querySelector('div.thf-page-bottom-container'));
            _this.toolbarElement = (document.querySelector('div.thf-toolbar'));
            _this.setHeightContent(_this.pageHeaderElement, _this.bottomActions, _this.toolbarElement);
        });
    };
    /**
     * @param {?} thfPageHeader
     * @param {?} bottomActions
     * @param {?} toolbarElement
     * @return {?}
     */
    ThfPageContentComponent.prototype.setHeightContent = function (thfPageHeader, bottomActions, toolbarElement) {
        var /** @type {?} */ toolbarElementHeight = toolbarElement ? 33 : 0;
        var /** @type {?} */ pageHeaderHeight = thfPageHeader ? thfPageHeader.offsetHeight : 0;
        var /** @type {?} */ bottomActionsHeight = bottomActions ? bottomActions.offsetHeight : 0;
        var /** @type {?} */ bodyHeight = document.body.clientHeight;
        var /** @type {?} */ newHeight = bodyHeight - (pageHeaderHeight + bottomActionsHeight + toolbarElementHeight);
        this.overflowY = bottomActions ? 'auto' : 'none';
        this.height = newHeight + "px";
        this.marginTop = pageHeaderHeight + toolbarElementHeight + "px";
    };
    ThfPageContentComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-page-content', template: "<div class=\"thf-page-content\" [style.height]=\"height\" [style.margin-top]=\"marginTop\" [style.overflow-y]=\"overflowY\"> <ng-content></ng-content> </div> "
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfPageContentComponent.ctorParameters = function () { return [
        { type: Renderer, },
    ]; };
    return ThfPageContentComponent;
}(ThfPageContentBaseComponent));
export { ThfPageContentComponent };
function ThfPageContentComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfPageContentComponent.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfPageContentComponent.ctorParameters;
    /** @type {?} */
    ThfPageContentComponent.prototype.height;
    /** @type {?} */
    ThfPageContentComponent.prototype.toolbarElement;
    /** @type {?} */
    ThfPageContentComponent.prototype.pageHeaderElement;
    /** @type {?} */
    ThfPageContentComponent.prototype.bottomActions;
    /** @type {?} */
    ThfPageContentComponent.prototype.marginTop;
    /** @type {?} */
    ThfPageContentComponent.prototype.overflowY;
    /** @type {?} */
    ThfPageContentComponent.prototype.renderer;
}
