// @totvs/thf-web/components/thf-tabs/thf-tab-button/thf-tab-button.component.js
import { Component, Input, Output, EventEmitter, ElementRef, ViewChild, HostListener } from '@angular/core';
var ThfTabButtonComponent = (function () {
    /**
     * @param {?} el
     */
    function ThfTabButtonComponent(el) {
        this.el = el;
        // Método recebido do usuário para ser disparado quando clicar na aba
        this.clickCallback = new EventEmitter;
        this.clickTab = new EventEmitter;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    ThfTabButtonComponent.prototype.onClick = function (e) {
        if (!this.disabled) {
            this.clickTab.emit(this.id);
            this.clickCallback.emit(this.id);
        }
    };
    /**
     * @return {?}
     */
    ThfTabButtonComponent.prototype.controlDisabled = function () {
        if (this.disabled) {
            this.el.nativeElement.classList.add('thf-tab-disabled');
        }
        else {
            this.el.nativeElement.classList.remove('thf-tab-disabled');
        }
    };
    /**
     * @return {?}
     */
    ThfTabButtonComponent.prototype.controlActive = function () {
        if (this.active) {
            this.el.nativeElement.classList.add('thf-tab-active');
        }
        else {
            this.el.nativeElement.classList.remove('thf-tab-active');
        }
    };
    /**
     * @return {?}
     */
    ThfTabButtonComponent.prototype.ngAfterViewChecked = function () {
        if (this.hide) {
            this.el.nativeElement.style.display = 'none';
        }
        else {
            this.el.nativeElement.style.display = '';
            this.el.nativeElement.style.width = this.width + '%';
        }
        this.controlDisabled();
        this.controlActive();
    };
    ThfTabButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-tab-button', template: "<div #tab [ngClass]=\"{'thf-tab-size1': !small, 'thf-tab-size2': small}\"> {{label}} </div> ", styles: [":host(thf-tab-button) { background-color: red; display: table-cell; background-color: transparent; box-shadow: inset 0 -1px 0 0 #edefef; text-align: center; vertical-align: middle; font-family: NunitoSans; font-weight: bold; color: #0c9abe; cursor: pointer; } :host(thf-tab-button) .thf-tab-size1 { display: inline-flex; align-items: center; text-align: center; height: 44px; font-size: 16px; } :host(thf-tab-button) .thf-tab-size2 { display: inline-flex; align-items: center; text-align: center; height: 32px; font-size: 14px; } :host(thf-tab-button) .thf-tab-hidden { display: none; } :host(thf-tab-button).thf-tab-disabled { color: rgba(0, 0, 0, 0.4); cursor: initial; } :host(thf-tab-button).thf-tab-active { box-shadow: inset 0 -3px 0 0 #29b6c5; } "]
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfTabButtonComponent.ctorParameters = function () { return [
        { type: ElementRef, },
    ]; };
    ThfTabButtonComponent.propDecorators = {
        'id': [{ type: Input },],
        'active': [{ type: Input },],
        'label': [{ type: Input },],
        'width': [{ type: Input },],
        'small': [{ type: Input },],
        'disabled': [{ type: Input },],
        'hide': [{ type: Input },],
        'clickCallback': [{ type: Output },],
        'clickTab': [{ type: Output },],
        'tab': [{ type: ViewChild, args: ['tab',] },],
        'onClick': [{ type: HostListener, args: ['click', ['$event'],] },],
    };
    return ThfTabButtonComponent;
}());
export { ThfTabButtonComponent };
function ThfTabButtonComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfTabButtonComponent.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfTabButtonComponent.ctorParameters;
    /** @type {?} */
    ThfTabButtonComponent.propDecorators;
    /** @type {?} */
    ThfTabButtonComponent.prototype.id;
    /** @type {?} */
    ThfTabButtonComponent.prototype.active;
    /** @type {?} */
    ThfTabButtonComponent.prototype.label;
    /** @type {?} */
    ThfTabButtonComponent.prototype.width;
    /** @type {?} */
    ThfTabButtonComponent.prototype.small;
    /** @type {?} */
    ThfTabButtonComponent.prototype.disabled;
    /** @type {?} */
    ThfTabButtonComponent.prototype.hide;
    /** @type {?} */
    ThfTabButtonComponent.prototype.clickCallback;
    /** @type {?} */
    ThfTabButtonComponent.prototype.clickTab;
    /** @type {?} */
    ThfTabButtonComponent.prototype.tab;
    /** @type {?} */
    ThfTabButtonComponent.prototype.el;
}
