// @totvs/thf-web/thf.module.js
import { NgModule } from '@angular/core';
import { ThfComponentsModule } from './components/components.module';
import { ThfDirectivesModule } from './directives/directives.module';
import { ThfGuardsModule } from './guards/guards.module';
import { ThfInterceptorsModule } from './interceptors/interceptors.module';
import { ThfPipesModule } from './pipes/pipes.module';
import { ThfServicesModule } from './services/services.module';
import { ThfNotificationService } from './services/thf-notification/thf-notification.service';
var ThfModule = (function () {
    function ThfModule() {
    }
    ThfModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        ThfComponentsModule,
                        ThfDirectivesModule,
                        ThfGuardsModule,
                        ThfInterceptorsModule,
                        ThfPipesModule,
                        ThfServicesModule
                    ],
                    exports: [
                        ThfComponentsModule,
                        ThfDirectivesModule,
                        ThfInterceptorsModule,
                        ThfGuardsModule,
                        ThfPipesModule,
                        ThfServicesModule
                    ],
                    providers: [ThfNotificationService],
                    bootstrap: []
                },] },
    ];
    /**
     * @nocollapse
     */
    ThfModule.ctorParameters = function () { return []; };
    return ThfModule;
}());
export { ThfModule };
function ThfModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ThfModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ThfModule.ctorParameters;
}
